---
title: "Working with HPAanalyze package"
author: "Anh N. Tran"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{"Working with HPAanalyze package}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)

library(HPAanalyze)
library(tibble)
library(dplyr)
```

## Summary
### Keywords

## Background
###Motivation:

HPAanalyze consists of

Previously, another R package `hpar` was developed to...

## Implementation

### The different HPA data formats

HPA provides downloadable data...

### `HPAanalyze` overview

#### Obtaining `HPAanalyze`

#### Primary functionality

### Working with HPA Downloadable Data



HPA analyze provide tools to retrive and analyze these data via a set of functions including...

#### Download and import data with `hpaDownload()`

```{r downloaded_data, warning=FALSE, message=FALSE, eval=FALSE}
downloaded_data <- hpaDownload(downloadList = 'all')
summary(downloaded_data)

#>                          Length Class  Mode
#> normal_tissue             6     tbl_df list
#> pathology                11     tbl_df list
#> subcellular_location     11     tbl_df list
#> rna_tissue                5     tbl_df list
#> rna_cell_line             5     tbl_df list
#> transcript_rna_tissue     4     tbl_df list
#> transcript_rna_cell_line  4     tbl_df list
```

The `normal_tissue` dataset contains information about protein expression profiles in human tissues based on IHC staining. The datasets contain six columns: `ensembl` (Ensembl gene identifier); `gene` (HGNC symbol), `tissue` (tissue name); `cell_type` (annotated cell type); `level` (expression value); `reliability` (the gene reliability of the expression value).

```{r normal_tissue}
downloaded_data <- hpaDownload(downloadList = 'histology', version = 'example')
tibble::glimpse(downloaded_data$normal_tissue, give.attr = FALSE)
```

The `pathology` dataset contains information about protein expression profiles in human tumor tissue based on IHC staining. The datasets contain eleven columns: `ensembl` (Ensembl gene identifier); `gene` (HGNC symbol); `cancer` (cancer type); `high`, `medium`, `low`, `not_detected` (number of patients annotated for different staining levels); `prognostic_favorable`, `unprognostic_favorable`, `prognostic_unfavorable`, `unprognostic_unfavorable` (log-rank p values for patient survival and mRNA correlation).

```{r pathology}
tibble::glimpse(downloaded_data$pathology, give.attr = FALSE)
```

The `subcellular_location` dataset contains information about subcellular localization of proteins based on IF stanings of normal cells. The datasets contain eleven columns: `ensembl` (Ensembl gene identifier); `gene` (HGNC symbol); `reliability` (gene reliability score); `enhanced` (enhanced locations); `supported` (supported locations); `approved` (approved locations); `uncertain` (uncertain locations); `single_cell_var_intensity` (locations with single-cell variation in intensity); `single_cell_var_spatial` (locations with spatial single-cell variation); `cell_cycle_dependency` (locations with observed cell cycle dependency); `go_id` (Gene Ontology Cellular Component term identifier).

```{r subcellular_location}
tibble::glimpse(downloaded_data$subcellular_location, give.attr = FALSE)
```

The `rna_tissue` and `rna_cell_line` datasets contain RNA expression levels of 37 tissues and 64 cell lines based on RNA-seq. These datasets contain four columns each: `ensembl` (Ensembl gene identifier); `gene` (HGNC symbol); `tissue`/`cell_line` (type of sample); `value` + `unit` (expression level measured by transcripts per million).

```{r rna, warning=FALSE, message=FALSE}
downloaded_data <- hpaDownload(downloadList = 'rna', version = 'hpar')

tibble::glimpse(downloaded_data$rna_tissue, give.attr = FALSE)

tibble::glimpse(downloaded_data$rna_cell_line, give.attr = FALSE)
```

Similarly, the `transcript_rna_tissue` and `transcript_rna_cell_line` datasets contain RNA isoform levels. These datasets contain four columns each: `ensembl` (Ensembl gene identifier); `transcript` (Ensembl transcript identifier); `tissue`/`cell_line` (type of sample); `value` (expression level measured by transcripts per million). Note that these datasets are significantly larger than others and should only be downloaded when necessary.

```{r eval=FALSE}
downloaded_data <- hpaDownload(downloadList = 'isoform', version = 'v18')

tibble::glimpse(downloaded_data$transcript_rna_tissue, give.attr = FALSE)

#> Observations: 27,535,996
#> Variables: 4
#> $ ensembl    <chr> "ENSG00000000003", "ENSG00000000003", "ENSG0000000...
#> $ transcript <chr> "ENST00000373020", "ENST00000494424", "ENST0000049...
#> $ tissue     <chr> "adipose tissue.V1", "adipose tissue.V1", "adipose...
#> $ value      <dbl> 27.3577003, 0.0000000, 1.9341500, 1.6059300, 0.000...

tibble::glimpse(downloaded_data$transcript_rna_cell_line, give.attr = FALSE)

#> Observations: 20,972,183
#> Variables: 4
#> $ ensembl    <chr> "ENSG00000000003", "ENSG00000000003", "ENSG0000000...
#> $ transcript <chr> "ENST00000373020", "ENST00000494424", "ENST0000049...
#> $ cell_line  <chr> "A-431.C35", "A-431.C35", "A-431.C35", "A-431.C35"...
#> $ value      <dbl> 29.406799, 0.000000, 0.992916, 0.398387, 0.239204,...
```

#### List available data with `hpaListParam()`

To see what parameters are available for subsequent subsetting/visualizing, HPAanalyze includes the function `hpa_list_param()`. The input for this function is the list object

```{r list_param, eval= FALSE}
downloaded_data <- hpaDownload(downloadList = 'all')
str(hpaListParam(downloaded_data))

#> List of 6
#>  $ normal_tissue       : chr [1:58] "adrenal gland" "appendix" "bone marrow" "breast" ...
#>  $ normal_cell         : chr [1:82] "glandular cells" "lymphoid tissue" "hematopoietic cells" "adipocytes" ...
#>  $ cancer              : chr [1:20] "breast cancer" "carcinoid" "cervical cancer" "colorectal cancer" ...
#>  $ subcellular_location: chr [1:32] "Cytosol" "Mitochondria" "Aggresome" "Plasma membrane" ...
#>  $ normal_tissue_rna   : chr [1:37] "adipose tissue" "adrenal gland" "appendix" "bone marrow" ...
#>  $ cell_line_rna       : chr [1:64] "A-431" "A549" "AF22" "AN3-CA" ...
```

#### Subset data with `hpaSubset()`

```{r subset, message=FALSE, warning=FALSE}
downloaded_data <- hpaDownload(downloadList = 'histology', version = 'example')
sapply(downloaded_data, nrow)

gene_list <- c('TP53', 'EGFR', 'CD44', 'PTEN', 'IDH1', 'IDH2', 'CYCS')
tissue_list <- c('breast', 'cerebellum', 'skin 1')
cancer_list <- c('breast cancer', 'glioma', 'melanoma')
cell_line_list <- c('A-431', 'A549', 'AF22', 'AN3-CA')

subset_data <- hpaSubset(data = downloaded_data,
                         targetGene = gene_list,
                         targetTissue = tissue_list,
                         targetCancer = cancer_list,
                         targetCellLine = cell_line_list)
sapply(subset_data, nrow)

```

#### Export data with `hpaExport()`

```{r eval=FALSE}
hpaExport(subset_data, fileName = 'subset.xlsx', fileType = 'xlsx')
```

### Visualize downloaded data

`HPAanalyze` provides the ability to quickly visualize data from downloaded HPA datasets with the `hpaVis` function family. The goal of these functions is to aid exploratory analysis of a group of target genes, which maybe particularly useful for gaining insights into pathways or gene signatures of interest.

The `hpaVis` functions share a common syntax, where the input (`data` argument) is the output of `hpaDownload()` or `hpaSubset()` (although they do their own subseting so it is not necessary to use `hpaSubset()` unless you want to reduce the size of your data object). Depending on the function, the `target` arguments will let you choose to visualize your vectors of genes, tissue, cell types, etc. (See the help files for more details.) All of `hpaVis` functions generate standard `ggplot2` plots, which allow you to further customize colors and themes. Colors maybe changed via the `color` argument, while the default theme maybe overriden by setting the `customTheme` argument to `FALSE`.

Currently, the `normal_tissue`, `pathology` and `subcellular_location` data can be visualized, with more functions planned for future releases.

```{r echo=FALSE, warning=FALSE, message=FALSE}
downloaded_data <- hpaDownload('histology', 'example')
```


#### Visualize tissue data with `hpaVisTissue()`

`hpaVisTissue()` generates a "heatmap", in which the expression of proteins of interest (quantified IHC staining) are plotted for each cell type of each tissue. 

```{r fig.width=6}
gene_list <- c('TP53', 'EGFR', 'CD44', 'PTEN', 'IDH1', 'IDH2', 'CYCS')
tissue_list <- c('breast', 'cerebellum', 'skin 1')

hpaVisTissue(downloaded_data,
             targetGene = gene_list,
             targetTissue = tissue_list)
```


#### Visualize expression in cancer with `hpaVisPatho()`

`hpaVisPatho()` generates an arrays of column graphs showing the expression of proteins of interest in each cancer. 

This example also demonstrate how the colors of the graphs could be customized, which is a common functionality of the `hpaVis` family.

```{r fig.width=6}
gene_list <- c('TP53', 'EGFR', 'CD44', 'PTEN', 'IDH1', 'IDH2', 'CYCS')
cancer_list <- c('breast cancer', 'glioma', 'lymphoma', 'prostate cancer')
color_gray <- c('slategray1', 'slategray2', 'slategray3', 'slategray4')

hpaVisPatho(downloaded_data,
            targetGene = gene_list,
            targetCancer = cancer_list,
            color = color_gray)
```


#### Visualize subcellular location data with `hpaVisSubcell()`

`hpaVisSubcell()` 

```{r fig.width=6}
gene_list <- c('TP53', 'EGFR', 'CD44', 'PTEN', 'IDH1', 'IDH2', 'CYCS')

hpaVisSubcell(downloaded_data,
              targetGene = gene_list)
```


### Working with individual xml files

#### Import xml file with `hpaXmlGet()`

```{r}
CCNB1_xml <- hpaXmlGet('ENSG00000134057')
```

#### View protein classes with `hpaXmlProtClass()`

```{r}
hpaXmlProtClass(CCNB1_xml)
```

#### Get summary and images of tissue expression with `hpaXmlTissueExprSum()`

```{r}
hpaXmlTissueExprSum(CCNB1_xml)
```

#### Get details of individual IHC samples with `hpaXmlAntibody()` and `hpaXmlTissueExpr()`

```{r}
hpaXmlAntibody(CCNB1_xml)

tissue_expression <- hpaXmlTissueExpr(CCNB1_xml)

summary(tissue_expression)

tissue_expression[[1]]

```

### Future developments

## Conclusion

## Availability and requirements

## Abbreviations

## Declarations

## References

## Copyright