<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>1. Quick-start guide: Acquire and visualize the Human Protein Atlas (HPA) data in one function with HPAanalyze â€¢ HPAanalyze</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="1. Quick-start guide: Acquire and visualize the Human Protein Atlas (HPA) data in one function with HPAanalyze">
<meta property="og:description" content="HPAanalyze">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">HPAanalyze</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.15.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/a_HPAanalyze_quick_start.html">1. Quick-start guide: Acquire and visualize the Human Protein Atlas (HPA) data in one function with HPAanalyze</a>
    </li>
    <li>
      <a href="../articles/b_HPAanalyze_indepth.html">2. In-depth: Working with Human Protein Atlas (HPA) data in R with HPAanalyze</a>
    </li>
    <li>
      <a href="../articles/c_HPAanalyze_case_query.html">3. Tutorial: Combine HPAanalyze with your Human Protein Atlas (HPA) queries</a>
    </li>
    <li>
      <a href="../articles/d_HPAanalyze_case_offline_xml.html">4. Tutorial: Working with Human Protein Atlas (HPA) xml files offline</a>
    </li>
    <li>
      <a href="../articles/e_HPAanalyze_case_json.html">5. Tutorial: Export Human Protein Atlas (HPA) data as JSON</a>
    </li>
    <li>
      <a href="../articles/f_HPAanalyze_case_images.html">6. Tutorial: Download histology images from the Human Protein Atlas</a>
    </li>
    <li>
      <a href="../articles/z_HPAanalyze_paper_figures.html">99. Code for figures from HPAanalyze paper</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/anhtr/HPAanalyze/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>1. Quick-start guide: Acquire and visualize the
Human Protein Atlas (HPA) data in one function with HPAanalyze</h1>
                        <h4 data-toc-skip class="author">Anh N.
Tran</h4>
            <address class="author_afil">
      DataGrata
LLC<br><a class="author_email" href="mailto:#"></a><a href="mailto:trannhatanh89@gmail.com" class="email">trannhatanh89@gmail.com</a>
      </address>
                  
            <h4 data-toc-skip class="date">4/1/2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/anhtr/HPAanalyze/blob/HEAD/vignettes/a_HPAanalyze_quick_start.Rmd" class="external-link"><code>vignettes/a_HPAanalyze_quick_start.Rmd</code></a></small>
      <div class="hidden name"><code>a_HPAanalyze_quick_start.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocStyle" class="external-link">BiocStyle</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/anhtr/HPAanalyze" class="external-link">HPAanalyze</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="background">Background<a class="anchor" aria-label="anchor" href="#background"></a>
</h2>
<p>The Human Protein Atlas (HPA) is a comprehensive resource for
exploration of human proteome which contains a vast amount of proteomics
and transcriptomics data generated from antibody-based tissue
micro-array profiling and RNA deep-sequencing.</p>
<p>The program has generated protein expression profiles in human normal
tissues with cell type-specific expression patterns, cancer and cell
lines via an innovative immunohistochemistry-based approach. These
profiles are accompanied by a large collection of high quality
histological staining images, annotated with clinical data and
quantification. The database also includes classification of protein
into both functional classes (such as transcription factors or kinases)
and project-related classes (such as candidate genes for cancer).
Starting from version 4.0, the HPA includes subcellular location
profiles generated based on confocal images of immunofluorescent stained
cells. Together, these data provide a detailed picture of protein
expression in human cells and tissues, facilitating tissue-based
diagnostic and research.</p>
<p>Data from the HPA are freely available via proteinatlas.org, allowing
scientists to access and incorporate the data into their research.
Previously, the R package <em>hpar</em> has been created for fast and
easy programmatic access of HPA data. Here, we introduce
<em>HPAanalyze</em>, an R package aims to simplify exploratory data
analysis from those data, as well as provide other complementary
functionality to <em>hpar</em>.</p>
<div class="section level3">
<h3 id="the-different-hpa-data-formats">The different HPA data formats<a class="anchor" aria-label="anchor" href="#the-different-hpa-data-formats"></a>
</h3>
<p>The Human Protein Atlas project provides data via two main
mechanisms: <em>Full datasets</em> in the form of downloadable
compressed tab-separated files (.tsv) and <em>individual entries</em> in
XML, RDF and TSV formats. The full downloadable datasets includes normal
tissue, pathology (cancer), subcellular location and RNA expression
data. For individual entries, the XML format is the most comprehensive,
providing information on the target protein, antibodies, summary for
each tissue and detailed data from each sample including clinical data,
IHC scoring and image download links.</p>
</div>
<div class="section level3">
<h3 id="hpaanalyze-overview">
<code>HPAanalyze</code> overview<a class="anchor" aria-label="anchor" href="#hpaanalyze-overview"></a>
</h3>
<p><code>HPAanalyze</code> is designed to fullfill 3 main tasks: (1)
Import, subsetting and export downloadable datasets; (2) Visualization
of downloadable datasets for exploratory analysis; and (3) Working with
the individual XML files. This package aims to serve researchers with
little programming experience, but also allow power users to use the
imported data as desired.</p>
</div>
</div>
<div class="section level2">
<h2 id="visualize-protein-expression-data">Visualize protein expression data<a class="anchor" aria-label="anchor" href="#visualize-protein-expression-data"></a>
</h2>
<p>Currently, this is available for the normal tissue, pathology
(cancers) and subcellular location datasets. The fastest and easiest way
is to use the defaults of <code>hpaVis</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/hpaVis.html">hpaVis</a></span><span class="op">(</span>targetGene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GCH1"</span>, <span class="st">"PTS"</span>, <span class="st">"SPR"</span>, <span class="st">"DHFR"</span><span class="op">)</span>,</span>
<span>       targetTissue <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cerebellum"</span>, <span class="st">"cerebral cortex"</span>, <span class="st">"hippocampus"</span><span class="op">)</span>,</span>
<span>       targetCancer <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"glioma"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; No data provided. Use version 21.1.</span></span>
<span><span class="co">#&gt; targetCellType variable not specified, visualize all.</span></span>
<span><span class="co">#&gt; * WARNING: targetCellType variable not specified, visualize all.</span></span>
<span><span class="co">#&gt; &gt;&gt; Use hpaListParam() to list possible values for target variables.</span></span>
<span><span class="co">#&gt; Use hpaListParam() to list possible values for target variables.</span></span></code></pre></div>
<p><img src="a_HPAanalyze_quick_start_files/figure-html/hpaVis_eg-1.png" width="700"></p>
<p>Of course, we cannot visualize everything in those big datasets, so
some defauts will be used and you will receive some warning
messages.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/hpaVis.html">hpaVis</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># No data provided. Use version ...</span></span>
<span><span class="co"># targetGene variable not specified, default to TP53, RB1, MYC, KRAS and EGFR.</span></span>
<span><span class="co"># targetTissue variable not specified, default to breast.</span></span>
<span><span class="co"># targetCellType variable not specified, visualize all.</span></span>
<span><span class="co"># * WARNING: targetCellType variable not specified, visualize all.</span></span>
<span><span class="co"># &gt;&gt; Use hpaListParam() to list possible values for target variables.</span></span>
<span><span class="co"># targetCancer variable not specified, default to breast cancer</span></span>
<span><span class="co"># Use hpaListParam() to list possible values for target variables.</span></span></code></pre></div>
<p>You can also use <code>hpaVis</code> to show just one or two of the
three graphs.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/hpaVis.html">hpaVis</a></span><span class="op">(</span>visType <span class="op">=</span> <span class="st">"Patho"</span>,</span>
<span>       targetGene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GCH1"</span>, <span class="st">"PTS"</span>, <span class="st">"SPR"</span>, <span class="st">"DHFR"</span><span class="op">)</span>,</span>
<span>       targetCancer <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"glioma"</span>, <span class="st">"breast cancer"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; No data provided. Use version 21.1.</span></span></code></pre></div>
<p><img src="a_HPAanalyze_quick_start_files/figure-html/hpaVis_eg3-1.png" width="700"></p>
<p>One exception though, if you want to plot all cancers, use
<code>hpaVisPatho</code> with <code>targetCancer = NULL</code>
(default).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/hpaVisPatho.html">hpaVisPatho</a></span><span class="op">(</span>targetGene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GCH1"</span>, <span class="st">"PTS"</span>, <span class="st">"SPR"</span>, <span class="st">"DHFR"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; No data provided. Use version 21.1.</span></span>
<span><span class="co">#&gt; * WARNING: targetCancer variable not specified, visualize all.</span></span>
<span><span class="co">#&gt; &gt;&gt; Use hpaListParam() to list possible values for target variables.</span></span></code></pre></div>
<p><img src="a_HPAanalyze_quick_start_files/figure-html/hpaVisPatho_eg-1.png" width="700"></p>
<p>There are many ways you could customize your plots. Please see the
documentation for more details.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">?</span><span class="va">hpaVis</span> <span class="co"># the easy umbrella to visualize protein expression levels</span></span>
<span><span class="op">?</span><span class="va">hpaVisTissue</span> <span class="co"># in normal tissue</span></span>
<span><span class="op">?</span><span class="va">hpaVisSubcell</span> <span class="co"># in subcellular compartments</span></span>
<span><span class="op">?</span><span class="va">hpaVisPatho</span> <span class="co"># in cancers</span></span></code></pre></div>
<p>If you want to know what kind of data you can visualize, use
<code>hpaListParam</code>. You will receive a list of parameters you can
use. Please note that if you ask the functions to plot anything that is
not on this list, they will just ignore it and plot what are
available.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/hpaListParam.html">hpaListParam</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; No data provided. Use version 21.1.</span></span>
<span><span class="co">#&gt; List of 2</span></span>
<span><span class="co">#&gt;  $ normal_tissue:List of 2</span></span>
<span><span class="co">#&gt;   ..$ tissue   : chr [1:63] "adipose tissue" "adrenal gland" "appendix" "bone marrow" ...</span></span>
<span><span class="co">#&gt;   ..$ cell_type: chr [1:141] "adipocytes" "glandular cells" "lymphoid tissue" "hematopoietic cells" ...</span></span>
<span><span class="co">#&gt;  $ pathology    :List of 1</span></span>
<span><span class="co">#&gt;   ..$ cancer: chr [1:20] "breast cancer" "carcinoid" "cervical cancer" "colorectal cancer" ...</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="acquiring-individual-sample-data-from-the-human-protein-atlas">Acquiring individual sample data from the Human Protein Atlas<a class="anchor" aria-label="anchor" href="#acquiring-individual-sample-data-from-the-human-protein-atlas"></a>
</h2>
<p>HPA provide data in two different formats: the more convenient
summarized data tables that are used for the <code>hpaVis</code>
functions above, and detailed annotated data for every sample and every
antibody in the xml format. There is one xml for each protein, which
contain all data generated by HPA. However, extracting information from
these xml files into a tidy format is a challenge. The
<code>hpaXml</code> functions are designed to help you easily access
this.</p>
<p>The easiest way is to use the umbrella <code><a href="../reference/hpaXml.html">hpaXml()</a></code>
function. Please note that at the moment this function only accept
Ensembl gene id.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">EGFR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hpaXml.html">hpaXml</a></span><span class="op">(</span>inputXml<span class="op">=</span><span class="st">'ENSG00000146648'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">EGFR</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#&gt; [1] "ProtClass"     "TissueExprSum" "Antibody"      "TissueExpr"   </span></span></code></pre></div>
<p>This function will return a list. The first item
<code>ProtClass</code> is the known and predicted class of the queried
protein. The second item <code>TissueExprSum</code> give you the quick
summary of protein expression in normal tissue and image url to download
one representative image. The third item <code>Antibody</code> provide
information on all antibodied used by HPA to stain for the queried
protein.</p>
<p>The last item <code>TissueExpr</code> is a list of data frame, one
for each of the antibodies. Each row will have everything available
regarding a sample stained with an antibody (clinical data, protein
expression via IHC and the Url to download the original image). If you
find an empty data frame, that means the antibody was used only for IF
staining to determine subcellular locations. Currently HPAanalyze does
not support extraction of that data.</p>
<p>To better understand the output, please read the documentation for
other <code>hpaXml</code> functions, which was called under the hood by
<code><a href="../reference/hpaXml.html">hpaXml()</a></code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">?</span><span class="va">hpaXmlGet</span> <span class="co"># import the xml file as "xml_document"</span></span>
<span><span class="op">?</span><span class="va">hpaXmlProtClass</span></span>
<span><span class="op">?</span><span class="va">hpaXmlTissueExprSum</span></span>
<span><span class="op">?</span><span class="va">hpaXmlAntibody</span></span>
<span><span class="op">?</span><span class="va">hpaXmlTissueExpr</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="copyright">Copyright<a class="anchor" aria-label="anchor" href="#copyright"></a>
</h2>
<p><strong>Anh Tran, 2018-2022</strong></p>
<p>Please cite: <strong>Tran, A.N., Dussaq, A.M., Kennell, T. et
al.Â HPAanalyze: an R package that facilitates the retrieval and analysis
of the Human Protein Atlas data. BMC Bioinformatics 20, 463 (2019) <a href="https://doi.org/10.1186/s12859-019-3059-z" class="external-link uri">https://doi.org/10.1186/s12859-019-3059-z</a></strong></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Anh Nhat Tran.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
