<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>2. HPAanalyze in-depth: Working with Human Protein Atlas (HPA) data in R • HPAanalyze</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="2. HPAanalyze in-depth: Working with Human Protein Atlas (HPA) data in R">
<meta property="og:description" content="HPAanalyze">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">HPAanalyze</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.8.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/a_HPAanalyze_quick_start.html">1. HPAanalyze quick start guide: Acquire and visualize the Human Protein Atlas (HPA) data in one function</a>
    </li>
    <li>
      <a href="../articles/b_HPAanalyze_indepth.html">2. HPAanalyze in-depth: Working with Human Protein Atlas (HPA) data in R</a>
    </li>
    <li>
      <a href="../articles/d_HPAanalyze_case_offline_xml.html">4. HPAanalyze use case: Working with Human Protein Atlas (HPA) xml files offline</a>
    </li>
    <li>
      <a href="../articles/c_HPAanalyze_case_query.html">3. HPAanalyze use case: Combine with your Human Protein Atlas (HPA) queries</a>
    </li>
    <li>
      <a href="../articles/e_HPAanalyze_case_json.html">5. HPAanalyze use case: Export Human Protein Atlas (HPA) data as JSON</a>
    </li>
    <li>
      <a href="../articles/f_HPAanalyze_case_images.html">6. HPAanalyze use case: Download histology images from the Human Protein Atlas</a>
    </li>
    <li>
      <a href="../articles/z_HPAanalyze_paper_figures.html">99. Code for figures from HPAanalyze paper</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/trannhatanh89/HPAanalyze/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="b_HPAanalyze_indepth_files/header-attrs-2.11/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>2. HPAanalyze in-depth: Working with Human Protein Atlas (HPA) data in R</h1>
                        <h4 class="author">Anh N. Tran</h4>
            <address class="author_afil">
      Northwestern University, IL, USA<br><a class="author_email" href="mailto:#"></a><a href="mailto:trannhatanh89@gmail.com" class="email">trannhatanh89@gmail.com</a>
      </address>
                  
            <h4 class="date">6/7/2019</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/trannhatanh89/HPAanalyze/blob/master/vignettes/b_HPAanalyze_indepth.Rmd"><code>vignettes/b_HPAanalyze_indepth.Rmd</code></a></small>
      <div class="hidden name"><code>b_HPAanalyze_indepth.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocStyle">BiocStyle</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/trannhatanh89/HPAanalyze">HPAanalyze</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/">tibble</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></code></pre></div>
<div id="summary" class="section level1">
<h1 class="hasAnchor">
<a href="#summary" class="anchor"></a>Summary</h1>
<ul>
<li>
<strong>Background:</strong> The Human Protein Atlas program aims to map human proteins via multiple technologies including imaging, proteomics and transcriptomics.</li>
<li>
<strong>Results:</strong> <code>HPAanalyze</code> is an R package for retreiving and performing exploratory data analysis from HPA. It provides functionality for importing data tables and xml files from HPA, exporting and visualizing data, as well as download all staining images of interest. The package is free, open source, and available via Github.</li>
<li>
<strong>Conclusions:</strong> <code>HPAanalyze</code> intergrates into the R workflow via the <code>tidyverse</code> philosophy and data structures, and can be used in combination with Bioconductor packages for easy analysis of HPA data.</li>
</ul>
<p><strong>Keywords:</strong> Human Protein Atlas, Proteomics, Homo Sapiens, Visualization, Software</p>
</div>
<div id="background" class="section level1">
<h1 class="hasAnchor">
<a href="#background" class="anchor"></a>Background</h1>
<p>The Human Protein Atlas (HPA) is a comprehensive resource for exploration of human proteome which contains a vast amount of proteomics and transcriptomics data generated from antibody-based tissue micro-array profiling and RNA deep-sequencing.</p>
<p>The program has generated protein expression profiles in human normal tissues with cell type-specific expression patterns, cancer and cell lines via an innovative immunohistochemistry-based approach. These profiles are accompanied by a large collection of high quality histological staining images, annotated with clinical data and quantification. The database also includes classification of protein into both functional classes (such as transcription factors or kinases) and project-related classes (such as candidate genes for cancer). Starting from version 4.0, the HPA includes subcellular location profiles generated based on confocal images of immunofluorescent stained cells. Together, these data provide a detailed picture of protein expression in human cells and tissues, facilitating tissue-based diagnostic and research.</p>
<p>Data from the HPA are freely available via proteinatlas.org, allowing scientists to access and incorporate the data into their research. Previously, the R package <em>hpar</em> has been created for fast and easy programmatic access of HPA data. Here, we introduce <em>HPAanalyze</em>, an R package aims to simplify exploratory data analysis from those data, as well as provide other complementary functionality to <em>hpar</em>.</p>
<div id="the-different-hpa-data-formats" class="section level2">
<h2 class="hasAnchor">
<a href="#the-different-hpa-data-formats" class="anchor"></a>The different HPA data formats</h2>
<p>The Human Protein Atlas project provides data via two main mechanisms: <em>Full datasets</em> in the form of downloadable compressed tab-separated files (.tsv) and <em>individual entries</em> in XML, RDF and TSV formats. The full downloadable datasets includes normal tissue, pathology (cancer), subcellular location, RNA gene and RNA isoform data. For individual entries, the XML format is the most comprehensive, providing information on the target protein, antibodies, summary for each tissue and detailed data from each sample including clinical data, IHC scoring and image download links.</p>
</div>
<div id="hpaanalyze-overview" class="section level2">
<h2 class="hasAnchor">
<a href="#hpaanalyze-overview" class="anchor"></a><code>HPAanalyze</code> overview</h2>
<p><code>HPAanalyze</code> is designed to fullfill 3 main tasks: (1) Import, subsetting and export downloadable datasets; (2) Visualization of downloadable datasets for exploratory analysis; and (3) Working with the individual XML files. This package aims to serve researchers with little programming experience, but also allow power users to use the imported data as desired.</p>
<div class="figure">
<img src="figures/workflow.png" alt="HPAanalyze workflow." width="100%"><p class="caption">
HPAanalyze workflow.
</p>
</div>
<div id="obtaining-hpaanalyze" class="section level3">
<h3 class="hasAnchor">
<a href="#obtaining-hpaanalyze" class="anchor"></a>Obtaining <code>HPAanalyze</code>
</h3>
<p>The stable version of <em>HPAanalyze</em> should be downloaded from Bioconductor:</p>
<pre><code>if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("HPAanalyze")</code></pre>
<p>The development version of <em>HPAanalyze</em> is available on Github can be installed with:</p>
<pre><code><a href="https://devtools.r-lib.org//reference/remote-reexports.html">devtools::install_github("trannhatanh89/HPAanalyze")</a></code></pre>
<p>Please cite: <strong>Tran AN, Dussaq AM, Kennell T, Willey C, Hjelmeland A. <em>HPAanalyze: An R Package that Facilitates the Retrieval and Analysis of The Human Protein Atlas Data</em>. bioRxiv 355032; doi: <a href="https://doi.org/10.1101/355032" class="uri">https://doi.org/10.1101/355032</a> </strong></p>
</div>
</div>
</div>
<div id="full-dataset-import-subsetting-and-export" class="section level1">
<h1 class="hasAnchor">
<a href="#full-dataset-import-subsetting-and-export" class="anchor"></a>Full dataset import, subsetting and export</h1>
<p>The <code><a href="../reference/hpaDownload.html">hpaDownload()</a></code> function downloads full datasets from HPA (<em>specifically, the .tsv format described above</em>) and imports them into R as a list of tibbles, the standard object of <code>tidyverse</code>, which can subsequently be subset with <code><a href="../reference/hpaSubset.html">hpaSubset()</a></code> and export into .xmlx files with <code><a href="../reference/hpaExport.html">hpaExport()</a></code>. The standard object allow the imported data to be further processed in a traditional R workflow. The ability to quickly subset and export data gives researchers the option to use other non-R downstream tools, such as GraphPad for creating publication-quality graphics, or share a subset of data containing only proteins of interest.</p>
<p><strong>You can skip this whole section if you only care about visualization, unless you need a specific version of the HPA datasets, or the RNA expression datasets.</strong></p>
<div id="download-and-import-data-with-hpadownload" class="section level2">
<h2 class="hasAnchor">
<a href="#download-and-import-data-with-hpadownload" class="anchor"></a>Download and import data with <code>hpaDownload()</code>
</h2>
<p>This function should be the first thing you use. It give you a list of data frames containing the datasets you specified, which you can then feed into other functions in this package.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># this gives you the latest everything, which is nice to keep but not really necessary</span>
<span class="va">downloadedData</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hpaDownload.html">hpaDownload</a></span><span class="op">(</span>downloadList<span class="op">=</span><span class="st">'all'</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">downloadedData</span><span class="op">)</span>

<span class="co">#&gt;                          Length Class  Mode</span>
<span class="co">#&gt; normal_tissue             6     tbl_df list</span>
<span class="co">#&gt; pathology                11     tbl_df list</span>
<span class="co">#&gt; subcellular_location     11     tbl_df list</span>
<span class="co">#&gt; rna_tissue                5     tbl_df list</span>
<span class="co">#&gt; rna_cell_line             5     tbl_df list</span>
<span class="co">#&gt; transcript_rna_tissue     4     tbl_df list</span>
<span class="co">#&gt; transcript_rna_cell_line  4     tbl_df list</span></code></pre></div>
<div id="the-histology-datasets" class="section level3">
<h3 class="hasAnchor">
<a href="#the-histology-datasets" class="anchor"></a>The “histology” datasets</h3>
<p>Most of the time, you will only need the “histology” datasets, which contain <code>normal_tissue</code>, <code>pathology</code> (basically cancers) and <code>subcellular_location</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">downloadedData</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hpaDownload.html">hpaDownload</a></span><span class="op">(</span>downloadList<span class="op">=</span><span class="st">'histology'</span>, version<span class="op">=</span><span class="st">'example'</span><span class="op">)</span>
<span class="co"># version = "example" will load the HPA v18 datasets came with this package. That's sufficient for normal usage, and save you some time.</span></code></pre></div>
<p>The <code>normal_tissue</code> dataset contains information about protein expression profiles in human tissues based on IHC staining. The datasets contain six columns: <code>ensembl</code> (Ensembl gene identifier); <code>gene</code> (HGNC symbol), <code>tissue</code> (tissue name); <code>cell_type</code> (annotated cell type); <code>level</code> (expression value); <code>reliability</code> (the gene reliability of the expression value).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html">glimpse</a></span><span class="op">(</span><span class="va">downloadedData</span><span class="op">$</span><span class="va">normal_tissue</span>, give.attr<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="co">#&gt; Rows: 1,118,517</span>
<span class="co">#&gt; Columns: 6</span>
<span class="co">#&gt; $ ensembl     &lt;chr&gt; "ENSG00000000003", "ENSG00000000003", "ENSG00000000003", "~</span>
<span class="co">#&gt; $ gene        &lt;chr&gt; "TSPAN6", "TSPAN6", "TSPAN6", "TSPAN6", "TSPAN6", "TSPAN6"~</span>
<span class="co">#&gt; $ tissue      &lt;chr&gt; "adipose tissue", "adrenal gland", "appendix", "appendix",~</span>
<span class="co">#&gt; $ cell_type   &lt;chr&gt; "adipocytes", "glandular cells", "glandular cells", "lymph~</span>
<span class="co">#&gt; $ level       &lt;chr&gt; "Not detected", "Not detected", "Medium", "Not detected", ~</span>
<span class="co">#&gt; $ reliability &lt;chr&gt; "Approved", "Approved", "Approved", "Approved", "Approved"~</span></code></pre></div>
<p>The <code>pathology</code> dataset contains information about protein expression profiles in human tumor tissue based on IHC staining. The datasets contain eleven columns: <code>ensembl</code> (Ensembl gene identifier); <code>gene</code> (HGNC symbol); <code>cancer</code> (cancer type); <code>high</code>, <code>medium</code>, <code>low</code>, <code>not_detected</code> (number of patients annotated for different staining levels); <code>prognostic_favorable</code>, <code>unprognostic_favorable</code>, <code>prognostic_unfavorable</code>, <code>unprognostic_unfavorable</code> (log-rank p values for patient survival and mRNA correlation).</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html">glimpse</a></span><span class="op">(</span><span class="va">downloadedData</span><span class="op">$</span><span class="va">pathology</span>, give.attr<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="co">#&gt; Rows: 393,400</span>
<span class="co">#&gt; Columns: 11</span>
<span class="co">#&gt; $ ensembl                  &lt;chr&gt; "ENSG00000000003", "ENSG00000000003", "ENSG00~</span>
<span class="co">#&gt; $ gene                     &lt;chr&gt; "TSPAN6", "TSPAN6", "TSPAN6", "TSPAN6", "TSPA~</span>
<span class="co">#&gt; $ cancer                   &lt;chr&gt; "breast cancer", "carcinoid", "cervical cance~</span>
<span class="co">#&gt; $ high                     &lt;int&gt; 1, 0, 11, 0, 10, 0, 0, 4, 8, 0, 0, 8, 11, 5, ~</span>
<span class="co">#&gt; $ medium                   &lt;int&gt; 7, 1, 1, 6, 2, 0, 3, 5, 4, 0, 1, 3, 1, 6, 4, ~</span>
<span class="co">#&gt; $ low                      &lt;int&gt; 2, 1, 0, 2, 0, 0, 1, 1, 0, 0, 2, 0, 0, 0, 1, ~</span>
<span class="co">#&gt; $ not_detected             &lt;int&gt; 2, 2, 0, 2, 0, 11, 0, 0, 0, 11, 9, 0, 0, 0, 5~</span>
<span class="co">#&gt; $ prognostic_favorable     &lt;chr&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N~</span>
<span class="co">#&gt; $ unprognostic_favorable   &lt;chr&gt; "7.712e-2", NA, "8.967e-2", "3.562e-2", NA, "~</span>
<span class="co">#&gt; $ prognostic_unfavorable   &lt;chr&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N~</span>
<span class="co">#&gt; $ unprognostic_unfavorable &lt;chr&gt; NA, NA, NA, NA, "2.567e-1", NA, "5.932e-2", N~</span></code></pre></div>
<p>The <code>subcellular_location</code> dataset contains information about subcellular localization of proteins based on IF stanings of normal cells. The datasets contain eleven columns: <code>ensembl</code> (Ensembl gene identifier); <code>gene</code> (HGNC symbol); <code>reliability</code> (gene reliability score); <code>enhanced</code> (enhanced locations); <code>supported</code> (supported locations); <code>approved</code> (approved locations); <code>uncertain</code> (uncertain locations); <code>single_cell_var_intensity</code> (locations with single-cell variation in intensity); <code>single_cell_var_spatial</code> (locations with spatial single-cell variation); <code>cell_cycle_dependency</code> (locations with observed cell cycle dependency); <code>go_id</code> (Gene Ontology Cellular Component term identifier).</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html">glimpse</a></span><span class="op">(</span><span class="va">downloadedData</span><span class="op">$</span><span class="va">subcellular_location</span>, give.attr<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="co">#&gt; Rows: 12,813</span>
<span class="co">#&gt; Columns: 14</span>
<span class="co">#&gt; $ ensembl                   &lt;chr&gt; "ENSG00000000003", "ENSG00000000457", "ENSG0~</span>
<span class="co">#&gt; $ gene                      &lt;chr&gt; "TSPAN6", "SCYL3", "C1orf112", "FGR", "CFH",~</span>
<span class="co">#&gt; $ reliability               &lt;chr&gt; "Approved", "Uncertain", "Approved", "Approv~</span>
<span class="co">#&gt; $ main_location             &lt;chr&gt; "Cell Junctions;Cytosol", "Microtubules", "M~</span>
<span class="co">#&gt; $ additional_location       &lt;chr&gt; "Nucleoli fibrillar center", "Nuclear bodies~</span>
<span class="co">#&gt; $ extracellular_location    &lt;chr&gt; NA, NA, NA, NA, "Predicted to be secreted", ~</span>
<span class="co">#&gt; $ enhanced                  &lt;chr&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, ~</span>
<span class="co">#&gt; $ supported                 &lt;chr&gt; NA, NA, NA, NA, NA, NA, "Nucleoplasm", NA, N~</span>
<span class="co">#&gt; $ approved                  &lt;chr&gt; "Cell Junctions;Cytosol;Nucleoli fibrillar c~</span>
<span class="co">#&gt; $ uncertain                 &lt;chr&gt; NA, "Microtubules;Nuclear bodies", NA, NA, N~</span>
<span class="co">#&gt; $ single_cell_var_intensity &lt;chr&gt; "Cytosol", NA, NA, NA, NA, "Cytosol;Nucleopl~</span>
<span class="co">#&gt; $ single_cell_var_spatial   &lt;chr&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, ~</span>
<span class="co">#&gt; $ cell_cycle_dependency     &lt;chr&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, ~</span>
<span class="co">#&gt; $ go_id                     &lt;chr&gt; "Cell Junctions (GO:0030054);Cytosol (GO:000~</span></code></pre></div>
</div>
<div id="the-rna-datasets" class="section level3">
<h3 class="hasAnchor">
<a href="#the-rna-datasets" class="anchor"></a>The “RNA” datasets</h3>
<p>The <code>rna_tissue</code> and <code>rna_cell_line</code> datasets contain RNA expression levels of 37 tissues and 64 cell lines based on RNA-seq. These datasets contain four columns each: <code>ensembl</code> (Ensembl gene identifier); <code>gene</code> (HGNC symbol); <code>tissue</code>/<code>cell_line</code> (type of sample); <code>value</code> + <code>unit</code> (expression level measured by transcripts per million).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">downloadedData</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hpaDownload.html">hpaDownload</a></span><span class="op">(</span>downloadList<span class="op">=</span><span class="st">'rna'</span>, version<span class="op">=</span><span class="st">'v18'</span><span class="op">)</span>

<span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html">glimpse</a></span><span class="op">(</span><span class="va">downloadedData</span><span class="op">$</span><span class="va">rna_tissue</span>, give.attr<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="co">#&gt;  NULL</span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html">glimpse</a></span><span class="op">(</span><span class="va">downloadedData</span><span class="op">$</span><span class="va">rna_cell_line</span>, give.attr<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="co">#&gt;  NULL</span></code></pre></div>
<p>Similarly, the <code>transcript_rna_tissue</code> and <code>transcript_rna_cell_line</code> datasets contain RNA isoform levels. These datasets contain four columns each: <code>ensembl</code> (Ensembl gene identifier); <code>transcript</code> (Ensembl transcript identifier); <code>tissue</code>/<code>cell_line</code> (type of sample); <code>value</code> (expression level measured by transcripts per million). Note that these datasets are significantly larger than others and should only be downloaded when necessary.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">downloadedData</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hpaDownload.html">hpaDownload</a></span><span class="op">(</span>downloadList<span class="op">=</span><span class="st">'isoform'</span>, version<span class="op">=</span><span class="st">'v18'</span><span class="op">)</span>
<span class="co"># version = "v18" is an example of how you may download different versions of the HPA datasets. Just change the number. Note that not all versions are available from the HPA website.</span>

<span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html">glimpse</a></span><span class="op">(</span><span class="va">downloadedData</span><span class="op">$</span><span class="va">transcript_rna_tissue</span>, give.attr<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>

<span class="co">#&gt; Observations: 27,535,996</span>
<span class="co">#&gt; Variables: 4</span>
<span class="co">#&gt; $ ensembl    &lt;chr&gt; "ENSG00000000003", "ENSG00000000003", "ENSG0000000...</span>
<span class="co">#&gt; $ transcript &lt;chr&gt; "ENST00000373020", "ENST00000494424", "ENST0000049...</span>
<span class="co">#&gt; $ tissue     &lt;chr&gt; "adipose tissue.V1", "adipose tissue.V1", "adipose...</span>
<span class="co">#&gt; $ value      &lt;dbl&gt; 27.3577003, 0.0000000, 1.9341500, 1.6059300, 0.000...</span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html">glimpse</a></span><span class="op">(</span><span class="va">downloadedData</span><span class="op">$</span><span class="va">transcript_rna_cell_line</span>, give.attr<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>

<span class="co">#&gt; Observations: 20,972,183</span>
<span class="co">#&gt; Variables: 4</span>
<span class="co">#&gt; $ ensembl    &lt;chr&gt; "ENSG00000000003", "ENSG00000000003", "ENSG0000000...</span>
<span class="co">#&gt; $ transcript &lt;chr&gt; "ENST00000373020", "ENST00000494424", "ENST0000049...</span>
<span class="co">#&gt; $ cell_line  &lt;chr&gt; "A-431.C35", "A-431.C35", "A-431.C35", "A-431.C35"...</span>
<span class="co">#&gt; $ value      &lt;dbl&gt; 29.406799, 0.000000, 0.992916, 0.398387, 0.239204,...</span></code></pre></div>
</div>
</div>
<div id="list-available-parameter-for-subsetting-with-hpalistparam" class="section level2">
<h2 class="hasAnchor">
<a href="#list-available-parameter-for-subsetting-with-hpalistparam" class="anchor"></a>List available parameter for subsetting with <code>hpaListParam()</code>
</h2>
<p>To see what parameters are available for subsequent subsetting/visualizing, HPAanalyze includes the function <code><a href="../reference/hpaSubset.html">hpaListParam()</a></code>. The input for this function is the output of <code>hpaDownload</code>.</p>
<p><em>If you leave the argument blank, this function will give you the results for version 18.</em></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## If you use the output from hpaDownload()</span>
<span class="va">downloadedData</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hpaDownload.html">hpaDownload</a></span><span class="op">(</span>downloadList<span class="op">=</span><span class="st">'all'</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="fu"><a href="../reference/hpaSubset.html">hpaListParam</a></span><span class="op">(</span><span class="va">downloadedData</span><span class="op">)</span><span class="op">)</span>

<span class="co">#&gt; List of 6</span>
<span class="co">#&gt;  $ normal_tissue       : chr [1:58] "adrenal gland" "appendix" "bone marrow" "breast" ...</span>
<span class="co">#&gt;  $ normal_cell         : chr [1:82] "glandular cells" "lymphoid tissue" "hematopoietic cells" "adipocytes" ...</span>
<span class="co">#&gt;  $ cancer              : chr [1:20] "breast cancer" "carcinoid" "cervical cancer" "colorectal cancer" ...</span>
<span class="co">#&gt;  $ subcellular_location: chr [1:32] "Cytosol" "Mitochondria" "Aggresome" "Plasma membrane" ...</span>
<span class="co">#&gt;  $ normal_tissue_rna   : chr [1:37] "adipose tissue" "adrenal gland" "appendix" "bone marrow" ...</span>
<span class="co">#&gt;  $ cell_line_rna       : chr [1:64] "A-431" "A549" "AF22" "AN3-CA" ...</span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## If you use leave the argument blank</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="fu"><a href="../reference/hpaSubset.html">hpaListParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; List of 4</span>
<span class="co">#&gt;  $ normal_tissue       : chr [1:63] "adipose tissue" "adrenal gland" "appendix" "bone marrow" ...</span>
<span class="co">#&gt;  $ normal_cell         : chr [1:120] "adipocytes" "glandular cells" "lymphoid tissue" "hematopoietic cells" ...</span>
<span class="co">#&gt;  $ cancer              : chr [1:20] "breast cancer" "carcinoid" "cervical cancer" "colorectal cancer" ...</span>
<span class="co">#&gt;  $ subcellular_location: chr [1:34] "Cell Junctions" "Cytosol" "Nucleoli fibrillar center" "Mitochondria" ...</span></code></pre></div>
</div>
<div id="subset-data-with-hpasubset" class="section level2">
<h2 class="hasAnchor">
<a href="#subset-data-with-hpasubset" class="anchor"></a>Subset data with <code>hpaSubset()</code>
</h2>
<p><code><a href="../reference/hpaSubset.html">hpaSubset()</a></code> filters the output of <code><a href="../reference/hpaDownload.html">hpaDownload()</a></code> for desirable target genes, tissues, cell types, cancer, and cell lines. The data will be subset only where applicable (i.e. <code>normal_tissue</code> will not be subset by <em>cancer</em>). The main purpose of <code>hpaSubset</code> is to prepare a manageable set of data to be exported. However, this function may also be useful for other data table manipulation purposes. The input for <code>targetGene</code> argument is a vector of strings of HGNC symbols.</p>
<p><em>If you leave the <code>data</code> argument blank, this function will automatically subset the bundled version 18 dataset, which may not contain all of the columns available if you download the data with <code><a href="../reference/hpaDownload.html">hpaDownload()</a></code>.</em></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">downloadedData</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hpaDownload.html">hpaDownload</a></span><span class="op">(</span>downloadList<span class="op">=</span><span class="st">'histology'</span>, version<span class="op">=</span><span class="st">'example'</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">downloadedData</span>, <span class="va">nrow</span><span class="op">)</span>
<span class="co">#&gt;        normal_tissue            pathology subcellular_location </span>
<span class="co">#&gt;              1118517               393400                12813</span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">geneList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'TP53'</span>, <span class="st">'EGFR'</span>, <span class="st">'CD44'</span>, <span class="st">'PTEN'</span>, <span class="st">'IDH1'</span>, <span class="st">'IDH2'</span>, <span class="st">'CYCS'</span><span class="op">)</span>
<span class="va">tissueList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'breast'</span>, <span class="st">'cerebellum'</span>, <span class="st">'skin 1'</span><span class="op">)</span>
<span class="va">cancerList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'breast cancer'</span>, <span class="st">'glioma'</span>, <span class="st">'melanoma'</span><span class="op">)</span>
<span class="va">cellLineList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'A-431'</span>, <span class="st">'A549'</span>, <span class="st">'AF22'</span>, <span class="st">'AN3-CA'</span><span class="op">)</span>

<span class="va">subsetData</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hpaSubset.html">hpaSubset</a></span><span class="op">(</span>data<span class="op">=</span><span class="va">downloadedData</span>,
                         targetGene<span class="op">=</span><span class="va">geneList</span>,
                         targetTissue<span class="op">=</span><span class="va">tissueList</span>,
                         targetCancer<span class="op">=</span><span class="va">cancerList</span>,
                         targetCellLine<span class="op">=</span><span class="va">cellLineList</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">subsetData</span>, <span class="va">nrow</span><span class="op">)</span>
<span class="co">#&gt;        normal_tissue            pathology subcellular_location </span>
<span class="co">#&gt;                   86                   21                    7</span></code></pre></div>
</div>
<div id="export-data-with-hpaexport" class="section level2">
<h2 class="hasAnchor">
<a href="#export-data-with-hpaexport" class="anchor"></a>Export data with <code>hpaExport()</code>
</h2>
<p>As the name suggests, <code><a href="../reference/hpaExport.html">hpaExport()</a></code> exports the output of <code><a href="../reference/hpaSubset.html">hpaSubset()</a></code> to one .xlsx file. Each dataset is placed in a separate sheet. More formats such as .csv and .tsv might be added in future release; hence, the <code>fileType</code> argument is included.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/hpaExport.html">hpaExport</a></span><span class="op">(</span><span class="va">subsetData</span>, fileName<span class="op">=</span><span class="st">'subset.xlsx'</span>, fileType<span class="op">=</span><span class="st">'xlsx'</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="visualization" class="section level1">
<h1 class="hasAnchor">
<a href="#visualization" class="anchor"></a>Visualization</h1>
<p><code>HPAanalyze</code> provides the ability to quickly visualize data from downloaded HPA datasets with the <code>hpaVis</code> function family. The goal of these functions is to aid exploratory analysis of a group of target genes, which maybe particularly useful for gaining insights into pathways or gene signatures of interest.</p>
<p>The <code>hpaVis</code> functions share a common syntax, where the input (<code>data</code> argument) is the output of <code><a href="../reference/hpaDownload.html">hpaDownload()</a></code> or <code><a href="../reference/hpaSubset.html">hpaSubset()</a></code> (although they do their own subseting so it is not necessary to use <code><a href="../reference/hpaSubset.html">hpaSubset()</a></code> unless you want to reduce the size of your data object). Depending on the function, the <code>target</code> arguments will let you choose to visualize your vectors of genes, tissue, cell types, etc. (See the help files for more details.) All of <code>hpaVis</code> functions generate standard <code>ggplot2</code> plots, which allow you to further customize colors and themes. Colors maybe changed via the <code>color</code> argument, while the default theme maybe overriden by setting the <code>customTheme</code> argument to <code>FALSE</code>.</p>
<p>Currently, the <code>normal_tissue</code>, <code>pathology</code> and <code>subcellular_location</code> data can be visualized, with more functions planned for future releases.</p>
<p><em>For all functions in the <code>hpaVis</code> family, if you leave the <code>data</code> argument blank, they will plot version 18 by default.</em></p>
<div id="unbrella-function-hpavis" class="section level2">
<h2 class="hasAnchor">
<a href="#unbrella-function-hpavis" class="anchor"></a>Unbrella function <code>hpaVis()</code>
</h2>
<p><code>hpaVis</code> will plot all available plots by default. See the quick-start vignette for details.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/hpaVis.html">hpaVis</a></span><span class="op">(</span>targetGene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"GCH1"</span>, <span class="st">"PTS"</span>, <span class="st">"SPR"</span>, <span class="st">"DHFR"</span><span class="op">)</span>,
       targetTissue <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"cerebellum"</span>, <span class="st">"cerebral cortex"</span>, <span class="st">"hippocampus"</span><span class="op">)</span>,
       targetCancer <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"glioma"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="b_HPAanalyze_indepth_files/figure-html/hpaVis_eg-1.png" width="700"></p>
</div>
<div id="visualize-tissue-data-with-hpavistissue" class="section level2">
<h2 class="hasAnchor">
<a href="#visualize-tissue-data-with-hpavistissue" class="anchor"></a>Visualize tissue data with <code>hpaVisTissue()</code>
</h2>
<p><code><a href="../reference/hpaVisTissue.html">hpaVisTissue()</a></code> generates a “heatmap”, in which the expression of proteins of interest (quantified IHC staining) are plotted for each cell type of each tissue.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">geneList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'TP53'</span>, <span class="st">'EGFR'</span>, <span class="st">'CD44'</span>, <span class="st">'PTEN'</span>, <span class="st">'IDH1'</span>, <span class="st">'IDH2'</span>, <span class="st">'CYCS'</span><span class="op">)</span>
<span class="va">tissueList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'breast'</span>, <span class="st">'cerebellum'</span>, <span class="st">'skin 1'</span><span class="op">)</span>

<span class="fu"><a href="../reference/hpaVisTissue.html">hpaVisTissue</a></span><span class="op">(</span><span class="va">downloadedData</span>,
             targetGene<span class="op">=</span><span class="va">geneList</span>,
             targetTissue<span class="op">=</span><span class="va">tissueList</span><span class="op">)</span></code></pre></div>
<p><img src="b_HPAanalyze_indepth_files/figure-html/visTissue-1.png" width="700"></p>
</div>
<div id="visualize-expression-in-cancer-with-hpavispatho" class="section level2">
<h2 class="hasAnchor">
<a href="#visualize-expression-in-cancer-with-hpavispatho" class="anchor"></a>Visualize expression in cancer with <code>hpaVisPatho()</code>
</h2>
<p><code><a href="../reference/hpaVisPatho.html">hpaVisPatho()</a></code> generates an arrays of column graphs showing the expression of proteins of interest in each cancer.</p>
<p>This example also demonstrate how the colors of the graphs could be customized, which is a common functionality of the <code>hpaVis</code> family.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">geneList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'TP53'</span>, <span class="st">'EGFR'</span>, <span class="st">'CD44'</span>, <span class="st">'PTEN'</span>, <span class="st">'IDH1'</span>, <span class="st">'IDH2'</span>, <span class="st">'CYCS'</span><span class="op">)</span>
<span class="va">cancerList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'breast cancer'</span>, <span class="st">'glioma'</span>, <span class="st">'lymphoma'</span>, <span class="st">'prostate cancer'</span><span class="op">)</span>
<span class="va">colorGray</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'slategray1'</span>, <span class="st">'slategray2'</span>, <span class="st">'slategray3'</span>, <span class="st">'slategray4'</span><span class="op">)</span>

<span class="fu"><a href="../reference/hpaVisPatho.html">hpaVisPatho</a></span><span class="op">(</span><span class="va">downloadedData</span>,
            targetGene<span class="op">=</span><span class="va">geneList</span>,
            targetCancer<span class="op">=</span><span class="va">cancerList</span>,
            color<span class="op">=</span><span class="va">colorGray</span><span class="op">)</span></code></pre></div>
<p><img src="b_HPAanalyze_indepth_files/figure-html/visPatho-1.png" width="700"></p>
</div>
<div id="visualize-subcellular-location-data-with-hpavissubcell" class="section level2">
<h2 class="hasAnchor">
<a href="#visualize-subcellular-location-data-with-hpavissubcell" class="anchor"></a>Visualize subcellular location data with <code>hpaVisSubcell()</code>
</h2>
<p><code><a href="../reference/hpaVisSubcell.html">hpaVisSubcell()</a></code> generates a tile chart showing the subcellular locations (approved and supported) of proteins of interest.</p>
<p>This example also demonstrate the customization of the output plot with ggplot2 functions, which is applicable to all <code>hpaVis</code> functions. Notice that the <code>customTheme</code> argument is set to <code>TRUE</code>.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">geneList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'TP53'</span>, <span class="st">'EGFR'</span>, <span class="st">'CD44'</span>, <span class="st">'PTEN'</span>, <span class="st">'IDH1'</span>, <span class="st">'IDH2'</span>, <span class="st">'CYCS'</span><span class="op">)</span>

<span class="fu"><a href="../reference/hpaVisSubcell.html">hpaVisSubcell</a></span><span class="op">(</span><span class="va">downloadedData</span>,
              targetGene<span class="op">=</span><span class="va">geneList</span>,
              customTheme<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">'Subcellular locations'</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">'Protein'</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle<span class="op">=</span><span class="fl">45</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span>  <span class="op">+</span>
    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position<span class="op">=</span><span class="st">"none"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html">coord_equal</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="b_HPAanalyze_indepth_files/figure-html/visSubcell-1.png" width="700"></p>
</div>
</div>
<div id="individual-xml-import-and-image-downloading" class="section level1">
<h1 class="hasAnchor">
<a href="#individual-xml-import-and-image-downloading" class="anchor"></a>Individual xml import and image downloading</h1>
<p>The <code>hpaXml</code> function family import and extract data from individual XML entries from HPA. The <code><a href="../reference/hpaXmlGet.html">hpaXmlGet()</a></code> function downloads and imports data as “xml_document”/“xml_node” object, which can subsequently be processed by other <code>hpaXml</code> functions. The XML format from HPA contains a wealth of information that may not be covered by this package. However, users can extract any data of interest from the imported XML file using the xml2 package.</p>
<p>A typical workflow for working with XML files includes the following steps: (1) Download and import XML file with <code><a href="../reference/hpaXmlGet.html">hpaXmlGet()</a></code>; (2) Extract the desired information with other <code>hpaXml</code> functions; and (3) Download histological staining pictures, which is currently supported by the <code>hpaXmlTissurExpr()</code> and <code><a href="../reference/hpaXmlTissueExprSum.html">hpaXmlTissueExprSum()</a></code> functions.</p>
<div id="the-umbrella-function-hpaxml" class="section level2">
<h2 class="hasAnchor">
<a href="#the-umbrella-function-hpaxml" class="anchor"></a>The umbrella function <code>hpaXml</code>
</h2>
<p><code>hpaXml</code> will take an Ensembl gene id (start with <em>ENSG</em>) and extract all availble information. You can also feed the ourput of hpaXmlGet to it. See the quick-start vignette for more details.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">EGFR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hpaXml.html">hpaXml</a></span><span class="op">(</span>inputXml<span class="op">=</span><span class="st">'ENSG00000146648'</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">EGFR</span><span class="op">)</span>

<span class="co">#&gt; [1] "ProtClass"     "TissueExprSum" "Antibody"      "TissueExpr"   </span></code></pre></div>
</div>
<div id="import-xml-file-with-hpaxmlget" class="section level2">
<h2 class="hasAnchor">
<a href="#import-xml-file-with-hpaxmlget" class="anchor"></a>Import xml file with <code>hpaXmlGet()</code>
</h2>
<p>The <code>hoaXmlGet()</code> function takes an Ensembl gene id (start with <em>ENSG</em>) and import the perspective XML file into R. This function calls the <code><a href="http://xml2.r-lib.org/reference/read_xml.html">xml2::read_xml()</a></code> under the hood, hence the resulting object may be processed further with <em>xml2</em> functions if desired.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">EGFRxml</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hpaXmlGet.html">hpaXmlGet</a></span><span class="op">(</span><span class="st">'ENSG00000146648'</span><span class="op">)</span></code></pre></div>
</div>
<div id="view-protein-classes-with-hpaxmlprotclass" class="section level2">
<h2 class="hasAnchor">
<a href="#view-protein-classes-with-hpaxmlprotclass" class="anchor"></a>View protein classes with <code>hpaXmlProtClass()</code>
</h2>
<p>Protein class of queried protein can be extracted from the imported XML with <code><a href="../reference/hpaXmlProtClass.html">hpaXmlProtClass()</a></code>. The output of this function is a tibble of 4 columns: <code>id</code>, <code>name</code>, <code>parent_id</code> and <code>source</code></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/hpaXmlProtClass.html">hpaXmlProtClass</a></span><span class="op">(</span><span class="va">EGFRxml</span><span class="op">)</span>

<span class="co">#&gt; # A tibble: 40 x 4</span>
<span class="co">#&gt;    id    name                                   parent_id source    </span>
<span class="co">#&gt;    &lt;chr&gt; &lt;chr&gt;                                  &lt;chr&gt;     &lt;chr&gt;     </span>
<span class="co">#&gt;  1 Ez    Enzymes                                &lt;NA&gt;      &lt;NA&gt;      </span>
<span class="co">#&gt;  2 Ec    ENZYME proteins                        Ez        ENZYME    </span>
<span class="co">#&gt;  3 Et    Transferases                           Ec        ENZYME    </span>
<span class="co">#&gt;  4 Ki    Kinases                                Ez        UniProt   </span>
<span class="co">#&gt;  5 Kt    Tyr protein kinases                    Ki        UniProt   </span>
<span class="co">#&gt;  6 Ma    Predicted membrane proteins            &lt;NA&gt;      MDM       </span>
<span class="co">#&gt;  7 Md    Membrane proteins predicted by MDM     &lt;NA&gt;      MDM       </span>
<span class="co">#&gt;  8 Me    MEMSAT3 predicted membrane proteins    &lt;NA&gt;      MEMSAT3   </span>
<span class="co">#&gt;  9 Mf    MEMSAT-SVM predicted membrane proteins &lt;NA&gt;      MEMSAT-SVM</span>
<span class="co">#&gt; 10 Mg    Phobius predicted membrane proteins    &lt;NA&gt;      Phobius   </span>
<span class="co">#&gt; # ... with 30 more rows</span></code></pre></div>
</div>
<div id="get-summary-and-images-of-tissue-expression-with-hpaxmltissueexprsum" class="section level2">
<h2 class="hasAnchor">
<a href="#get-summary-and-images-of-tissue-expression-with-hpaxmltissueexprsum" class="anchor"></a>Get summary and images of tissue expression with <code>hpaXmlTissueExprSum()</code>
</h2>
<p>The function <code><a href="../reference/hpaXmlTissueExprSum.html">hpaXmlTissueExprSum()</a></code> extract the summary of expression of protein of interest in normal tissue. The output of this function is a list of (1) a string contains one-sentence summary and (2) a dataframe of all tissues in which the protein was stained positive and a histological stain images of those tissue.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/hpaXmlTissueExprSum.html">hpaXmlTissueExprSum</a></span><span class="op">(</span><span class="va">EGFRxml</span><span class="op">)</span>

<span class="co">#&gt; $summary</span>
<span class="co">#&gt; [1] "Cytoplasmic and membranous expression in several tissues, most abundant in placenta."</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $img</span>
<span class="co">#&gt;            tissue</span>
<span class="co">#&gt; 1 cerebral cortex</span>
<span class="co">#&gt; 2      lymph node</span>
<span class="co">#&gt; 3           liver</span>
<span class="co">#&gt; 4           colon</span>
<span class="co">#&gt; 5          kidney</span>
<span class="co">#&gt; 6          testis</span>
<span class="co">#&gt; 7        placenta</span>
<span class="co">#&gt;                                                                imageUrl</span>
<span class="co">#&gt; 1 http://v18.proteinatlas.org/images/18530/41191_B_7_5_rna_selected.jpg</span>
<span class="co">#&gt; 2 http://v18.proteinatlas.org/images/18530/41191_A_7_8_rna_selected.jpg</span>
<span class="co">#&gt; 3 http://v18.proteinatlas.org/images/18530/41191_A_7_4_rna_selected.jpg</span>
<span class="co">#&gt; 4 http://v18.proteinatlas.org/images/18530/41191_A_9_3_rna_selected.jpg</span>
<span class="co">#&gt; 5 http://v18.proteinatlas.org/images/18530/41191_A_9_5_rna_selected.jpg</span>
<span class="co">#&gt; 6 http://v18.proteinatlas.org/images/18530/41191_A_6_6_rna_selected.jpg</span>
<span class="co">#&gt; 7 http://v18.proteinatlas.org/images/18530/41191_A_1_7_rna_selected.jpg</span></code></pre></div>
<p>Those images can be downloaded automatically by setting the <code>downloadImg</code> argument to <code>TRUE</code>. Eg. <code><a href="../reference/hpaXmlTissueExprSum.html">hpaXmlTissueExprSum(CCNB1xml, downloadImg=TRUE)</a></code></p>
</div>
<div id="get-details-of-individual-ihc-samples-with-hpaxmlantibody-and-hpaxmltissueexpr" class="section level2">
<h2 class="hasAnchor">
<a href="#get-details-of-individual-ihc-samples-with-hpaxmlantibody-and-hpaxmltissueexpr" class="anchor"></a>Get details of individual IHC samples with <code>hpaXmlAntibody()</code> and <code>hpaXmlTissueExpr()</code>
</h2>
<p>More importantly, the XML files are the only format of HPA programmatically accesible data which contains information about each antibody and each tissue sample used in the project.</p>
<p><code><a href="../reference/hpaXmlAntibody.html">hpaXmlAntibody()</a></code> extract the antibody information and return a tibble with one row for each antibody.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/hpaXmlAntibody.html">hpaXmlAntibody</a></span><span class="op">(</span><span class="va">EGFRxml</span><span class="op">)</span>

<span class="co">#&gt; # A tibble: 5 x 4</span>
<span class="co">#&gt;   id        releaseDate releaseVersion RRID      </span>
<span class="co">#&gt;   &lt;chr&gt;     &lt;chr&gt;       &lt;chr&gt;          &lt;chr&gt;     </span>
<span class="co">#&gt; 1 CAB000035 2006-03-13  1.2            &lt;NA&gt;      </span>
<span class="co">#&gt; 2 HPA001200 2008-02-15  3.1            AB_1078723</span>
<span class="co">#&gt; 3 HPA018530 2008-12-03  4.1            AB_1848044</span>
<span class="co">#&gt; 4 CAB068186 2014-11-06  13             AB_2665679</span>
<span class="co">#&gt; 5 CAB073534 2015-10-16  14             &lt;NA&gt;</span></code></pre></div>
<p><code><a href="../reference/hpaXmlTissueExpr.html">hpaXmlTissueExpr()</a></code> extract information about all samples for each antibody above and return a list of tibbles. If antibody has not been used for IHC staining, the returned tibble with be empty.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tissueExpression</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hpaXmlTissueExpr.html">hpaXmlTissueExpr</a></span><span class="op">(</span><span class="va">EGFRxml</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">tissueExpression</span><span class="op">)</span>

<span class="co">#&gt;      Length Class  Mode</span>
<span class="co">#&gt; [1,] 18     tbl_df list</span>
<span class="co">#&gt; [2,] 18     tbl_df list</span>
<span class="co">#&gt; [3,] 18     tbl_df list</span>
<span class="co">#&gt; [4,] 18     tbl_df list</span>
<span class="co">#&gt; [5,] 18     tbl_df list</span></code></pre></div>
<p>Each tibble contain clinical data (<code>patientid</code>, <code>age</code>, <code>sex</code>), tissue information (<code>snomedCode</code>, <code>tissueDescription</code>), staining results (<code>staining</code>, <code>intensity</code>, <code>location</code>) and one <code>imageUrl</code> for each sample. However, due to the large amount of data and the relatively large size of each image, <code>hpaXmlTissueExpr</code> does not provide an automated download option.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tissueExpression</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>

<span class="co">#&gt; # A tibble: 327 x 18</span>
<span class="co">#&gt;    patientId age   sex   staining intensity quantity location imageUrl</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;    &lt;chr&gt;     &lt;chr&gt;    &lt;chr&gt;    &lt;chr&gt;   </span>
<span class="co">#&gt;  1 1653      53    Male  &lt;NA&gt;     &lt;NA&gt;      &lt;NA&gt;     &lt;NA&gt;     http://~</span>
<span class="co">#&gt;  2 1721      60    Fema~ &lt;NA&gt;     &lt;NA&gt;      &lt;NA&gt;     &lt;NA&gt;     http://~</span>
<span class="co">#&gt;  3 1725      57    Male  &lt;NA&gt;     &lt;NA&gt;      &lt;NA&gt;     &lt;NA&gt;     http://~</span>
<span class="co">#&gt;  4 4         25    Male  &lt;NA&gt;     &lt;NA&gt;      &lt;NA&gt;     &lt;NA&gt;     http://~</span>
<span class="co">#&gt;  5 512       34    Fema~ &lt;NA&gt;     &lt;NA&gt;      &lt;NA&gt;     &lt;NA&gt;     http://~</span>
<span class="co">#&gt;  6 2664      74    Fema~ &lt;NA&gt;     &lt;NA&gt;      &lt;NA&gt;     &lt;NA&gt;     http://~</span>
<span class="co">#&gt;  7 2665      88    Fema~ &lt;NA&gt;     &lt;NA&gt;      &lt;NA&gt;     &lt;NA&gt;     http://~</span>
<span class="co">#&gt;  8 1391      54    Fema~ &lt;NA&gt;     &lt;NA&gt;      &lt;NA&gt;     &lt;NA&gt;     http://~</span>
<span class="co">#&gt;  9 1447      45    Fema~ &lt;NA&gt;     &lt;NA&gt;      &lt;NA&gt;     &lt;NA&gt;     http://~</span>
<span class="co">#&gt; 10 1452      44    Fema~ &lt;NA&gt;     &lt;NA&gt;      &lt;NA&gt;     &lt;NA&gt;     http://~</span>
<span class="co">#&gt; # ... with 317 more rows, and 10 more variables: snomedCode1 &lt;chr&gt;,</span>
<span class="co">#&gt; #   snomedCode2 &lt;chr&gt;, snomedCode3 &lt;chr&gt;, snomedCode4 &lt;chr&gt;,</span>
<span class="co">#&gt; #   snomedCode5 &lt;chr&gt;, tissueDescription1 &lt;chr&gt;, tissueDescription2 &lt;chr&gt;,</span>
<span class="co">#&gt; #   tissueDescription3 &lt;chr&gt;, tissueDescription4 &lt;chr&gt;,</span>
<span class="co">#&gt; #   tissueDescription5 &lt;chr&gt;</span></code></pre></div>
<p><code>hpaTissueExprSum</code> and <code>hpaTissueExpr</code> provide download links to download relevant staining images, with the former function also gives the options to automate the downloading process.</p>
</div>
</div>
<div id="compatibility-with-hpar-bioconductor-package" class="section level1">
<h1 class="hasAnchor">
<a href="#compatibility-with-hpar-bioconductor-package" class="anchor"></a>Compatibility with <code>hpar</code> Bioconductor package</h1>
<table class="table">
<caption>(#tab:table) Complementary functionality between hpar and HPAanalyze</caption>
<colgroup>
<col width="16%">
<col width="41%">
<col width="41%">
</colgroup>
<thead><tr class="header">
<th align="right">Functionality</th>
<th align="left">hpar</th>
<th align="left">HPAanalyze</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">Datasets</td>
<td align="left">Included in package</td>
<td align="left">Download from server or use built-in dataset</td>
</tr>
<tr class="even">
<td align="right">Query</td>
<td align="left">Ensembl id</td>
<td align="left">HGNC symbol for datasets, Ensembl id for XML</td>
</tr>
<tr class="odd">
<td align="right">Data version</td>
<td align="left">One stable version</td>
<td align="left">Latest by default, option to download older</td>
</tr>
<tr class="even">
<td align="right">Release info</td>
<td align="left">Access via functions</td>
<td align="left">N/A</td>
</tr>
<tr class="odd">
<td align="right">View relevant browser page</td>
<td align="left">Via <code>getHPA</code> function</td>
<td align="left">N/A</td>
</tr>
<tr class="even">
<td align="right">Visualization</td>
<td align="left">N/A</td>
<td align="left">Exploratory via <code>hpaVis</code> functions</td>
</tr>
<tr class="odd">
<td align="right">XML</td>
<td align="left">N/A</td>
<td align="left">Download and import via <code>hpaXml</code> functions</td>
</tr>
<tr class="even">
<td align="right">Histology image</td>
<td align="left">View by loading browser page</td>
<td align="left">Extract links via <code>hpaXml</code> functions</td>
</tr>
</tbody>
</table>
</div>
<div id="acknowledgements" class="section level1">
<h1 class="hasAnchor">
<a href="#acknowledgements" class="anchor"></a>Acknowledgements</h1>
<p>We appreciate the support of the National institutes of Health National Cancer Institute R01 CA151522 and funds from the Department of Cell, Developmental and Integrative Biology at the University of Alabama at Birmingham.</p>
</div>
<div id="copyright" class="section level1">
<h1 class="hasAnchor">
<a href="#copyright" class="anchor"></a>Copyright</h1>
<p><strong>Anh Tran, 2018-2021 </strong></p>
<p>Please cite: <strong>Tran, A.N., Dussaq, A.M., Kennell, T. et al. HPAanalyze: an R package that facilitates the retrieval and analysis of the Human Protein Atlas data. BMC Bioinformatics 20, 463 (2019) <a href="https://doi.org/10.1186/s12859-019-3059-z" class="uri">https://doi.org/10.1186/s12859-019-3059-z</a> </strong></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Anh Nhat Tran.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
