<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>2. In-depth: Working with Human Protein Atlas (HPA) data in R with HPAanalyze • HPAanalyze</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="2. In-depth: Working with Human Protein Atlas (HPA) data in R with HPAanalyze">
<meta property="og:description" content="HPAanalyze">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">HPAanalyze</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.15.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/a_HPAanalyze_quick_start.html">1. Quick-start guide: Acquire and visualize the Human Protein Atlas (HPA) data in one function with HPAanalyze</a>
    </li>
    <li>
      <a href="../articles/b_HPAanalyze_indepth.html">2. In-depth: Working with Human Protein Atlas (HPA) data in R with HPAanalyze</a>
    </li>
    <li>
      <a href="../articles/c_HPAanalyze_case_query.html">3. Tutorial: Combine HPAanalyze with your Human Protein Atlas (HPA) queries</a>
    </li>
    <li>
      <a href="../articles/d_HPAanalyze_case_offline_xml.html">4. Tutorial: Working with Human Protein Atlas (HPA) xml files offline</a>
    </li>
    <li>
      <a href="../articles/e_HPAanalyze_case_json.html">5. Tutorial: Export Human Protein Atlas (HPA) data as JSON</a>
    </li>
    <li>
      <a href="../articles/f_HPAanalyze_case_images.html">6. Tutorial: Download histology images from the Human Protein Atlas</a>
    </li>
    <li>
      <a href="../articles/z_HPAanalyze_paper_figures.html">99. Code for figures from HPAanalyze paper</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/anhtr/HPAanalyze/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>2. In-depth: Working with Human Protein Atlas
(HPA) data in R with HPAanalyze</h1>
                        <h4 data-toc-skip class="author">Anh N.
Tran</h4>
            <address class="author_afil">
      DataGrata
LLC<br><a class="author_email" href="mailto:#"></a><a href="mailto:trannhatanh89@gmail.com" class="email">trannhatanh89@gmail.com</a>
      </address>
                  
            <h4 data-toc-skip class="date">4/2/2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/anhtr/HPAanalyze/blob/HEAD/vignettes/b_HPAanalyze_indepth.Rmd" class="external-link"><code>vignettes/b_HPAanalyze_indepth.Rmd</code></a></small>
      <div class="hidden name"><code>b_HPAanalyze_indepth.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocStyle" class="external-link">BiocStyle</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/anhtr/HPAanalyze" class="external-link">HPAanalyze</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/" class="external-link">tibble</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<ul>
<li>
<strong>Background:</strong> The Human Protein Atlas program aims to
map human proteins via multiple technologies including imaging,
proteomics and transcriptomics.</li>
<li>
<strong>Results:</strong> <code>HPAanalyze</code> is an R package
for retreiving and performing exploratory data analysis from HPA. It
provides functionality for importing data tables and xml files from HPA,
exporting and visualizing data, as well as download all staining images
of interest. The package is free, open source, and available via
Github.</li>
<li>
<strong>Conclusions:</strong> <code>HPAanalyze</code> intergrates
into the R workflow via the <code>tidyverse</code> philosophy and data
structures, and can be used in combination with Bioconductor packages
for easy analysis of HPA data.</li>
</ul>
<p><strong>Keywords:</strong> Human Protein Atlas, Proteomics, Homo
Sapiens, Visualization, Software</p>
</div>
<div class="section level2">
<h2 id="background">Background<a class="anchor" aria-label="anchor" href="#background"></a>
</h2>
<p>The Human Protein Atlas (HPA) is a comprehensive resource for
exploration of human proteome which contains a vast amount of proteomics
and transcriptomics data generated from antibody-based tissue
micro-array profiling and RNA deep-sequencing.</p>
<p>The program has generated protein expression profiles in human normal
tissues with cell type-specific expression patterns, cancer and cell
lines via an innovative immunohistochemistry-based approach. These
profiles are accompanied by a large collection of high quality
histological staining images, annotated with clinical data and
quantification. The database also includes classification of protein
into both functional classes (such as transcription factors or kinases)
and project-related classes (such as candidate genes for cancer).
Starting from version 4.0, the HPA includes subcellular location
profiles generated based on confocal images of immunofluorescent stained
cells. Together, these data provide a detailed picture of protein
expression in human cells and tissues, facilitating tissue-based
diagnostic and research.</p>
<p>Data from the HPA are freely available via proteinatlas.org, allowing
scientists to access and incorporate the data into their research.
Previously, the R package <em>hpar</em> has been created for fast and
easy programmatic access of HPA data. Here, we introduce
<em>HPAanalyze</em>, an R package aims to simplify exploratory data
analysis from those data, as well as provide other complementary
functionality to <em>hpar</em>.</p>
<div class="section level3">
<h3 id="the-different-hpa-data-formats">The different HPA data formats<a class="anchor" aria-label="anchor" href="#the-different-hpa-data-formats"></a>
</h3>
<p>The Human Protein Atlas project provides data via two main
mechanisms: <em>Full datasets</em> in the form of downloadable
compressed tab-separated files (.tsv) and <em>individual entries</em> in
XML, RDF and TSV formats. The full downloadable datasets includes normal
tissue, pathology (cancer), subcellular location and RNA expression
data. For individual entries, the XML format is the most comprehensive,
providing information on the target protein, antibodies, summary for
each tissue and detailed data from each sample including clinical data,
IHC scoring and image download links.</p>
</div>
<div class="section level3">
<h3 id="hpaanalyze-overview">
<code>HPAanalyze</code> overview<a class="anchor" aria-label="anchor" href="#hpaanalyze-overview"></a>
</h3>
<p><code>HPAanalyze</code> is designed to fullfill 3 main tasks: (1)
Import, subsetting and export downloadable datasets; (2) Visualization
of downloadable datasets for exploratory analysis; and (3) Working with
the individual XML files. This package aims to serve researchers with
little programming experience, but also allow power users to use the
imported data as desired.</p>
<div class="figure">
<img src="figures/workflow.png" alt="HPAanalyze workflow." width="100%"><p class="caption">
HPAanalyze workflow.
</p>
</div>
<div class="section level4">
<h4 id="obtaining-hpaanalyze">Obtaining <code>HPAanalyze</code><a class="anchor" aria-label="anchor" href="#obtaining-hpaanalyze"></a>
</h4>
<p>The stable version of <em>HPAanalyze</em> should be downloaded from
Bioconductor:</p>
<pre><code><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"HPAanalyze"</span><span class="op">)</span></span></code></pre>
<p>The development version of <em>HPAanalyze</em> is available on Github
can be installed with:</p>
<pre><code><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"anhtr/HPAanalyze"</span><span class="op">)</span></span></code></pre>
<p>Please cite: <strong>Tran AN, Dussaq AM, Kennell T, Willey C,
Hjelmeland A. <em>HPAanalyze: An R Package that Facilitates the
Retrieval and Analysis of The Human Protein Atlas Data</em>. bioRxiv
355032; doi: <a href="https://doi.org/10.1101/355032" class="external-link uri">https://doi.org/10.1101/355032</a> </strong></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="full-dataset-import-subsetting-and-export">Full dataset import, subsetting and export<a class="anchor" aria-label="anchor" href="#full-dataset-import-subsetting-and-export"></a>
</h2>
<p>The <code><a href="../reference/hpaDownload.html">hpaDownload()</a></code> function downloads full datasets from
HPA (<em>specifically, the .tsv format described above</em>) and imports
them into R as a list of tibbles, the standard object of
<code>tidyverse</code>, which can subsequently be subset with
<code><a href="../reference/hpaListParam.html">hpaSubset()</a></code> and export into .xlsx files with
<code><a href="../reference/hpaExport.html">hpaExport()</a></code>. The standard object allow the imported data to
be further processed in a traditional R workflow. The ability to quickly
subset and export data gives researchers the option to use other non-R
downstream tools, such as GraphPad for creating publication-quality
graphics, or share a subset of data containing only proteins of
interest.</p>
<p><strong>You can skip this whole section if you only care about
visualization, unless you need a specific version of the HPA datasets,
or the RNA expression datasets.</strong></p>
<div class="section level3">
<h3 id="download-and-import-data-with-hpadownload">Download and import data with <code>hpaDownload()</code><a class="anchor" aria-label="anchor" href="#download-and-import-data-with-hpadownload"></a>
</h3>
<p>This function should be the first thing you use. It give you a list
of data frames containing the datasets you specified, which you can then
feed into other functions in this package.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># This should give you the latest of everything, but unless you have a lot of RAM and processing power, I would not recomend it.</span></span>
<span><span class="va">downloadedData</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hpaDownload.html">hpaDownload</a></span><span class="op">(</span>downloadList<span class="op">=</span><span class="st">'all'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">downloadedData</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#&gt;                          Length Class  Mode</span></span>
<span><span class="co">#&gt; normal_tissue             6     tbl_df list</span></span>
<span><span class="co">#&gt; pathology                11     tbl_df list</span></span>
<span><span class="co">#&gt; subcellular_location     14     tbl_df list</span></span>
<span><span class="co">#&gt; ...</span></span></code></pre></div>
<div class="section level4">
<h4 id="the-histology-datasets">The “histology” datasets<a class="anchor" aria-label="anchor" href="#the-histology-datasets"></a>
</h4>
<p>Most of the time, you will only need the “histology” datasets, which
contain <code>normal_tissue</code>, <code>pathology</code> (basically
cancers) and <code>subcellular_location</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">downloadedData</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hpaDownload.html">hpaDownload</a></span><span class="op">(</span>downloadList<span class="op">=</span><span class="st">'histology'</span>, version<span class="op">=</span><span class="st">'example'</span><span class="op">)</span></span>
<span><span class="co"># version = "example" will load the built-in dataset. That's sufficient for normal usage, and save you some time.</span></span></code></pre></div>
<p>The <code>normal_tissue</code> dataset contains information about
protein expression profiles in human tissues based on IHC staining. The
datasets contain six columns: <code>ensembl</code> (Ensembl gene
identifier); <code>gene</code> (HGNC symbol), <code>tissue</code>
(tissue name); <code>cell_type</code> (annotated cell type);
<code>level</code> (expression value); <code>reliability</code> (the
gene reliability of the expression value).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="va">downloadedData</span><span class="op">$</span><span class="va">normal_tissue</span>, give.attr<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Rows: 1,193,218</span></span>
<span><span class="co">#&gt; Columns: 6</span></span>
<span><span class="co">#&gt; $ ensembl     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "ENSG00000000003", "ENSG00000000003", "ENSG00000000003", "~</span></span>
<span><span class="co">#&gt; $ gene        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "TSPAN6", "TSPAN6", "TSPAN6", "TSPAN6", "TSPAN6", "TSPAN6"~</span></span>
<span><span class="co">#&gt; $ tissue      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "adipose tissue", "adrenal gland", "appendix", "appendix",~</span></span>
<span><span class="co">#&gt; $ cell_type   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "adipocytes", "glandular cells", "glandular cells", "lymph~</span></span>
<span><span class="co">#&gt; $ level       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "Not detected", "Not detected", "Medium", "Not detected", ~</span></span>
<span><span class="co">#&gt; $ reliability <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "Approved", "Approved", "Approved", "Approved", "Approved"~</span></span></code></pre></div>
<p>The <code>pathology</code> dataset contains information about protein
expression profiles in human tumor tissue based on IHC staining. The
datasets contain eleven columns: <code>ensembl</code> (Ensembl gene
identifier); <code>gene</code> (HGNC symbol); <code>cancer</code>
(cancer type); <code>high</code>, <code>medium</code>, <code>low</code>,
<code>not_detected</code> (number of patients annotated for different
staining levels); <code>prognostic_favorable</code>,
<code>unprognostic_favorable</code>,
<code>prognostic_unfavorable</code>,
<code>unprognostic_unfavorable</code> (log-rank p values for patient
survival and mRNA correlation).</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="va">downloadedData</span><span class="op">$</span><span class="va">pathology</span>, give.attr<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Rows: 401,800</span></span>
<span><span class="co">#&gt; Columns: 11</span></span>
<span><span class="co">#&gt; $ ensembl                  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "ENSG00000000003", "ENSG00000000003", "ENSG00~</span></span>
<span><span class="co">#&gt; $ gene                     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "TSPAN6", "TSPAN6", "TSPAN6", "TSPAN6", "TSPA~</span></span>
<span><span class="co">#&gt; $ cancer                   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "breast cancer", "carcinoid", "cervical cance~</span></span>
<span><span class="co">#&gt; $ high                     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> 1, 0, 11, 0, 10, 0, 0, 4, 8, 0, 0, 8, 11, 5, ~</span></span>
<span><span class="co">#&gt; $ medium                   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> 7, 1, 1, 6, 2, 0, 3, 5, 4, 0, 1, 3, 1, 6, 4, ~</span></span>
<span><span class="co">#&gt; $ low                      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> 2, 1, 0, 2, 0, 0, 1, 1, 0, 0, 2, 0, 0, 0, 1, ~</span></span>
<span><span class="co">#&gt; $ not_detected             <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> 2, 2, 0, 2, 0, 11, 0, 0, 0, 11, 9, 0, 0, 0, 5~</span></span>
<span><span class="co">#&gt; $ prognostic_favorable     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N~</span></span>
<span><span class="co">#&gt; $ unprognostic_favorable   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "7.712e-2", NA, "8.967e-2", "3.562e-2", NA, "~</span></span>
<span><span class="co">#&gt; $ prognostic_unfavorable   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N~</span></span>
<span><span class="co">#&gt; $ unprognostic_unfavorable <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> NA, NA, NA, NA, "2.567e-1", NA, "5.932e-2", N~</span></span></code></pre></div>
<p>The <code>subcellular_location</code> dataset contains information
about subcellular localization of proteins based on IF stanings of
normal cells. The datasets contain eleven columns: <code>ensembl</code>
(Ensembl gene identifier); <code>gene</code> (HGNC symbol);
<code>reliability</code> (gene reliability score); <code>enhanced</code>
(enhanced locations); <code>supported</code> (supported locations);
<code>approved</code> (approved locations); <code>uncertain</code>
(uncertain locations); <code>single_cell_var_intensity</code> (locations
with single-cell variation in intensity);
<code>single_cell_var_spatial</code> (locations with spatial single-cell
variation); <code>cell_cycle_dependency</code> (locations with observed
cell cycle dependency); <code>go_id</code> (Gene Ontology Cellular
Component term identifier).</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="va">downloadedData</span><span class="op">$</span><span class="va">subcellular_location</span>, give.attr<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Rows: 13,041</span></span>
<span><span class="co">#&gt; Columns: 14</span></span>
<span><span class="co">#&gt; $ ensembl                   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "ENSG00000000003", "ENSG00000000457", "ENSG0~</span></span>
<span><span class="co">#&gt; $ gene                      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "TSPAN6", "SCYL3", "C1orf112", "FGR", "CFH",~</span></span>
<span><span class="co">#&gt; $ reliability               <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "Approved", "Uncertain", "Approved", "Approv~</span></span>
<span><span class="co">#&gt; $ main_location             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "Cell Junctions;Cytosol", "Microtubules", "M~</span></span>
<span><span class="co">#&gt; $ additional_location       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "Nucleoli fibrillar center", "Nuclear bodies~</span></span>
<span><span class="co">#&gt; $ extracellular_location    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> NA, NA, NA, NA, "Predicted to be secreted", ~</span></span>
<span><span class="co">#&gt; $ enhanced                  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, ~</span></span>
<span><span class="co">#&gt; $ supported                 <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> NA, NA, NA, NA, NA, NA, "Nucleoplasm", NA, N~</span></span>
<span><span class="co">#&gt; $ approved                  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "Cell Junctions;Cytosol;Nucleoli fibrillar c~</span></span>
<span><span class="co">#&gt; $ uncertain                 <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> NA, "Microtubules;Nuclear bodies", NA, NA, N~</span></span>
<span><span class="co">#&gt; $ single_cell_var_intensity <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "Cytosol", NA, NA, NA, NA, "Cytosol;Nucleopl~</span></span>
<span><span class="co">#&gt; $ single_cell_var_spatial   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, ~</span></span>
<span><span class="co">#&gt; $ cell_cycle_dependency     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, ~</span></span>
<span><span class="co">#&gt; $ go_id                     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "Cell Junctions (GO:0030054);Cytosol (GO:000~</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="the-rna-datasets">The RNA datasets<a class="anchor" aria-label="anchor" href="#the-rna-datasets"></a>
</h4>
<p>HPA provide a large number of RNA expression datasets. More details
can be find at <a href="https://www.proteinatlas.org/about/download" class="external-link uri">https://www.proteinatlas.org/about/download</a>. To see how
to download these datasets, please see the help page for
<code>hpaDownload</code>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">?</span><span class="va">hpaDownload</span></span></code></pre></div>
<p>There are shortcuts to download multiple RNA datasets, for
example:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">downloadedData</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hpaDownload.html">hpaDownload</a></span><span class="op">(</span>downloadList<span class="op">=</span><span class="st">'rna tissue'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="va">downloadedData</span>, give.attr<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#&gt; List of 4</span></span>
<span><span class="co">#&gt;  $ rna_tissue_consensus: tibble [1,180,464 x 4] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="co">#&gt;   ..$ ensembl: chr [1:1180464] "ENSG00000000003" "ENSG00000000003" "ENSG00000000003" "ENSG00000000003" ...</span></span>
<span><span class="co">#&gt;   ..$ gene   : chr [1:1180464] "TSPAN6" "TSPAN6" "TSPAN6" "TSPAN6" ...</span></span>
<span><span class="co">#&gt;   ..$ tissue : chr [1:1180464] "adipose tissue" "adrenal gland" "amygdala" "appendix" ...</span></span>
<span><span class="co">#&gt;   ..$ nx     : chr [1:1180464] "27.0" "9.8" "7.0" "4.4" ...</span></span>
<span><span class="co">#&gt;  $ rna_tissue_hpa      : tibble [845,810 x 6] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="co">#&gt;   ..$ ensembl: chr [1:845810] "ENSG00000000003" "ENSG00000000003" "ENSG00000000003" "ENSG00000000003" ...</span></span>
<span><span class="co">#&gt;   ..$ gene   : chr [1:845810] "TSPAN6" "TSPAN6" "TSPAN6" "TSPAN6" ...</span></span>
<span><span class="co">#&gt;   ..$ tissue : chr [1:845810] "adipose tissue" "adrenal gland" "appendix" "B-cells" ...</span></span>
<span><span class="co">#&gt;   ..$ tpm    : chr [1:845810] "31.5" "26.4" "9.2" "0.1" ...</span></span>
<span><span class="co">#&gt;   ..$ ptpm   : chr [1:845810] "37.7" "32.7" "14.5" "0.2" ...</span></span>
<span><span class="co">#&gt;   ..$ nx     : chr [1:845810] "9.8" "7.6" "2.1" "0.3" ...</span></span>
<span><span class="co">#&gt;  $ rna_tissue_gtex     : tibble [639,744 x 6] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="co">#&gt;   ..$ ensembl: chr [1:639744] "ENSG00000000003" "ENSG00000000003" "ENSG00000000003" "ENSG00000000003" ...</span></span>
<span><span class="co">#&gt;   ..$ gene   : chr [1:639744] "TSPAN6" "TSPAN6" "TSPAN6" "TSPAN6" ...</span></span>
<span><span class="co">#&gt;   ..$ tissue : chr [1:639744] "adipose tissue" "adrenal gland" "amygdala" "basal ganglia" ...</span></span>
<span><span class="co">#&gt;   ..$ tpm    : chr [1:639744] "27.4" "15.5" "7.3" "7.7" ...</span></span>
<span><span class="co">#&gt;   ..$ ptpm   : chr [1:639744] "34.2" "18.8" "9.0" "9.4" ...</span></span>
<span><span class="co">#&gt;   ..$ nx     : chr [1:639744] "13.4" "9.8" "7.0" "6.6" ...</span></span>
<span><span class="co">#&gt;  $ rna_tissue_fantom   : tibble [797,265 x 6] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="co">#&gt;   ..$ ensembl                : chr [1:797265] "ENSG00000000003" "ENSG00000000003" "ENSG00000000003" "ENSG00000000003" ...</span></span>
<span><span class="co">#&gt;   ..$ gene                   : chr [1:797265] "TSPAN6" "TSPAN6" "TSPAN6" "TSPAN6" ...</span></span>
<span><span class="co">#&gt;   ..$ tissue                 : chr [1:797265] "adipose tissue" "amygdala" "appendix" "basal ganglia" ...</span></span>
<span><span class="co">#&gt;   ..$ tags_per_million       : chr [1:797265] "60.1" "10.7" "13.5" "16.2" ...</span></span>
<span><span class="co">#&gt;   ..$ scaled_tags_per_million: chr [1:797265] "87.4" "12.4" "17.8" "18.0" ...</span></span>
<span><span class="co">#&gt;   ..$ nx                     : chr [1:797265] "27.0" "3.5" "4.4" "5.5" ...</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="list-available-parameter-for-subsetting-with-hpalistparam">List available parameter for subsetting with
<code>hpaListParam()</code><a class="anchor" aria-label="anchor" href="#list-available-parameter-for-subsetting-with-hpalistparam"></a>
</h3>
<p>To see what parameters are available for subsequent
subsetting/visualizing, HPAanalyze includes the function
<code><a href="../reference/hpaListParam.html">hpaListParam()</a></code>. The input for this function is the output
of <code>hpaDownload</code>.</p>
<p><em>If you leave the argument blank, this function will give you the
results for the built-in datasets.</em></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## If you use the output from hpaDownload()</span></span>
<span><span class="va">downloadedData</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hpaDownload.html">hpaDownload</a></span><span class="op">(</span>downloadList<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Normal tissue"</span>, <span class="st">"Pathology"</span>, <span class="st">"RNA HPA tissue"</span>, <span class="st">"RNA HPA cell line"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="fu"><a href="../reference/hpaListParam.html">hpaListParam</a></span><span class="op">(</span><span class="va">downloadedData</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># List of 4</span></span>
<span><span class="co">#  $ normal_tissue :List of 2</span></span>
<span><span class="co">#   ..$ tissue   : chr [1:63] "adipose tissue" "adrenal gland" "appendix" "bone marrow" ...</span></span>
<span><span class="co">#   ..$ cell_type: chr [1:120] "adipocytes" "glandular cells" "lymphoid tissue" "hematopoietic cells" ...</span></span>
<span><span class="co">#  $ pathology     :List of 1</span></span>
<span><span class="co">#   ..$ cancer: chr [1:20] "breast cancer" "carcinoid" "cervical cancer" "colorectal cancer" ...</span></span>
<span><span class="co">#  $ rna_tissue_hpa:List of 1</span></span>
<span><span class="co">#   ..$ tissue: chr [1:43] "adipose tissue" "adrenal gland" "appendix" "B-cells" ...</span></span>
<span><span class="co">#  $ rna_celline   :List of 1</span></span>
<span><span class="co">#   ..$ cell_line: chr [1:69] "A-431" "A549" "AF22" "AN3-CA" ...</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="subset-data-with-hpasubset">Subset data with <code>hpaSubset()</code><a class="anchor" aria-label="anchor" href="#subset-data-with-hpasubset"></a>
</h3>
<p><code><a href="../reference/hpaListParam.html">hpaSubset()</a></code> filters the output of
<code><a href="../reference/hpaDownload.html">hpaDownload()</a></code> for desirable target genes, tissues, cell
types, cancer, and cell lines. The data will be subset only where
applicable (i.e. <code>normal_tissue</code> will not be subset by
<em>cancer</em>). The main purpose of <code>hpaSubset</code> is to
prepare a manageable set of data to be exported. However, this function
may also be useful for other data table manipulation purposes. The input
for <code>targetGene</code> argument is a vector of strings of HGNC
symbols.</p>
<p><em>If you leave the <code>data</code> argument blank, this function
will automatically subset the built-in dataset, which may not contain
all of the columns available if you download the data with
<code><a href="../reference/hpaDownload.html">hpaDownload()</a></code>.</em></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">downloadedData</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hpaDownload.html">hpaDownload</a></span><span class="op">(</span>downloadList<span class="op">=</span><span class="st">'histology'</span>, version<span class="op">=</span><span class="st">'example'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">downloadedData</span>, <span class="va">nrow</span><span class="op">)</span></span>
<span><span class="co">#&gt;        normal_tissue            pathology subcellular_location </span></span>
<span><span class="co">#&gt;              1193218               401800                13041</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">geneList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'TP53'</span>, <span class="st">'EGFR'</span>, <span class="st">'CD44'</span>, <span class="st">'PTEN'</span>, <span class="st">'IDH1'</span>, <span class="st">'IDH2'</span>, <span class="st">'CYCS'</span><span class="op">)</span></span>
<span><span class="va">tissueList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'breast'</span>, <span class="st">'cerebellum'</span>, <span class="st">'skin 1'</span><span class="op">)</span></span>
<span><span class="va">cancerList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'breast cancer'</span>, <span class="st">'glioma'</span>, <span class="st">'melanoma'</span><span class="op">)</span></span>
<span><span class="va">cellLineList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'A-431'</span>, <span class="st">'A549'</span>, <span class="st">'AF22'</span>, <span class="st">'AN3-CA'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">subsetData</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hpaListParam.html">hpaSubset</a></span><span class="op">(</span>data<span class="op">=</span><span class="va">downloadedData</span>,</span>
<span>                         targetGene<span class="op">=</span><span class="va">geneList</span>,</span>
<span>                         targetTissue<span class="op">=</span><span class="va">tissueList</span>,</span>
<span>                         targetCancer<span class="op">=</span><span class="va">cancerList</span>,</span>
<span>                         targetCellLine<span class="op">=</span><span class="va">cellLineList</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">subsetData</span>, <span class="va">nrow</span><span class="op">)</span></span>
<span><span class="co">#&gt;        normal_tissue            pathology subcellular_location </span></span>
<span><span class="co">#&gt;                  126                   21                    7</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="export-data-with-hpaexport">Export data with <code>hpaExport()</code><a class="anchor" aria-label="anchor" href="#export-data-with-hpaexport"></a>
</h3>
<p>As the name suggests, <code><a href="../reference/hpaExport.html">hpaExport()</a></code> exports the output of
<code><a href="../reference/hpaListParam.html">hpaSubset()</a></code> to one .xlsx file, or multiple .csv or .tsv
files. Each dataset is placed in a separate sheet.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/hpaExport.html">hpaExport</a></span><span class="op">(</span><span class="va">subsetData</span>, fileName<span class="op">=</span><span class="st">'subset.xlsx'</span>, fileType<span class="op">=</span><span class="st">'xlsx'</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h2>
<p><code>HPAanalyze</code> provides the ability to quickly visualize
data from downloaded HPA datasets with the <code>hpaVis</code> function
family. The goal of these functions is to aid exploratory analysis of a
group of target genes, which maybe particularly useful for gaining
insights into pathways or gene signatures of interest.</p>
<p>The <code>hpaVis</code> functions share a common syntax, where the
input (<code>data</code> argument) is the output of
<code><a href="../reference/hpaDownload.html">hpaDownload()</a></code> or <code><a href="../reference/hpaListParam.html">hpaSubset()</a></code> (although they do
their own subseting so it is not necessary to use
<code><a href="../reference/hpaListParam.html">hpaSubset()</a></code> unless you want to reduce the size of your data
object). Depending on the function, the <code>target</code> arguments
will let you choose to visualize your vectors of genes, tissue, cell
types, etc. (See the help files for more details.) All of
<code>hpaVis</code> functions generate standard <code>ggplot2</code>
plots, which allow you to further customize colors and themes. Colors
maybe changed via the <code>color</code> argument, while the default
theme maybe overriden by setting the <code>customTheme</code> argument
to <code>FALSE</code>.</p>
<p>Currently, the <code>normal_tissue</code>, <code>pathology</code> and
<code>subcellular_location</code> data can be visualized, with more
functions planned for future releases.</p>
<p><em>For all functions in the <code>hpaVis</code> family, if you leave
the <code>data</code> argument blank, they will plot the built-in
datasets by default.</em></p>
<div class="section level3">
<h3 id="unbrella-function-hpavis">Unbrella function <code>hpaVis()</code><a class="anchor" aria-label="anchor" href="#unbrella-function-hpavis"></a>
</h3>
<p><code>hpaVis</code> will plot all available plots by default. See the
quick-start vignette for details.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/hpaVis.html">hpaVis</a></span><span class="op">(</span><span class="va">downloadedData</span>,</span>
<span>       targetGene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GCH1"</span>, <span class="st">"PTS"</span>, <span class="st">"SPR"</span>, <span class="st">"DHFR"</span><span class="op">)</span>,</span>
<span>       targetTissue <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cerebellum"</span>, <span class="st">"cerebral cortex"</span>, <span class="st">"hippocampus"</span><span class="op">)</span>,</span>
<span>       targetCancer <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"glioma"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="b_HPAanalyze_indepth_files/figure-html/hpaVis_eg-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="visualize-tissue-data-with-hpavistissue">Visualize tissue data with <code>hpaVisTissue()</code><a class="anchor" aria-label="anchor" href="#visualize-tissue-data-with-hpavistissue"></a>
</h3>
<p><code><a href="../reference/hpaVisTissue.html">hpaVisTissue()</a></code> generates a “heatmap”, in which the
expression of proteins of interest (quantified IHC staining) are plotted
for each cell type of each tissue.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">geneList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'TP53'</span>, <span class="st">'EGFR'</span>, <span class="st">'CD44'</span>, <span class="st">'PTEN'</span>, <span class="st">'IDH1'</span>, <span class="st">'IDH2'</span>, <span class="st">'CYCS'</span><span class="op">)</span></span>
<span><span class="va">tissueList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'breast'</span>, <span class="st">'cerebellum'</span>, <span class="st">'skin 1'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/hpaVisTissue.html">hpaVisTissue</a></span><span class="op">(</span><span class="va">downloadedData</span>,</span>
<span>             targetGene<span class="op">=</span><span class="va">geneList</span>,</span>
<span>             targetTissue<span class="op">=</span><span class="va">tissueList</span><span class="op">)</span></span></code></pre></div>
<p><img src="b_HPAanalyze_indepth_files/figure-html/visTissue-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="visualize-expression-in-cancer-with-hpavispatho">Visualize expression in cancer with <code>hpaVisPatho()</code><a class="anchor" aria-label="anchor" href="#visualize-expression-in-cancer-with-hpavispatho"></a>
</h3>
<p><code><a href="../reference/hpaVisPatho.html">hpaVisPatho()</a></code> generates an arrays of column graphs
showing the expression of proteins of interest in each cancer.</p>
<p>This example also demonstrate how the colors of the graphs could be
customized, which is a common functionality of the <code>hpaVis</code>
family.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">geneList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'TP53'</span>, <span class="st">'EGFR'</span>, <span class="st">'CD44'</span>, <span class="st">'PTEN'</span>, <span class="st">'IDH1'</span>, <span class="st">'IDH2'</span>, <span class="st">'CYCS'</span><span class="op">)</span></span>
<span><span class="va">cancerList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'breast cancer'</span>, <span class="st">'glioma'</span>, <span class="st">'lymphoma'</span>, <span class="st">'prostate cancer'</span><span class="op">)</span></span>
<span><span class="va">colorGray</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'slategray1'</span>, <span class="st">'slategray2'</span>, <span class="st">'slategray3'</span>, <span class="st">'slategray4'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/hpaVisPatho.html">hpaVisPatho</a></span><span class="op">(</span><span class="va">downloadedData</span>,</span>
<span>            targetGene<span class="op">=</span><span class="va">geneList</span>,</span>
<span>            targetCancer<span class="op">=</span><span class="va">cancerList</span>,</span>
<span>            color<span class="op">=</span><span class="va">colorGray</span><span class="op">)</span></span></code></pre></div>
<p><img src="b_HPAanalyze_indepth_files/figure-html/visPatho-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="visualize-subcellular-location-data-with-hpavissubcell">Visualize subcellular location data with
<code>hpaVisSubcell()</code><a class="anchor" aria-label="anchor" href="#visualize-subcellular-location-data-with-hpavissubcell"></a>
</h3>
<p><code><a href="../reference/hpaVisSubcell.html">hpaVisSubcell()</a></code> generates a tile chart showing the
subcellular locations (approved and supported) of proteins of
interest.</p>
<p>This example also demonstrate the customization of the output plot
with ggplot2 functions, which is applicable to all <code>hpaVis</code>
functions. Notice that the <code>customTheme</code> argument is set to
<code>TRUE</code>.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">geneList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'TP53'</span>, <span class="st">'EGFR'</span>, <span class="st">'CD44'</span>, <span class="st">'PTEN'</span>, <span class="st">'IDH1'</span>, <span class="st">'IDH2'</span>, <span class="st">'CYCS'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/hpaVisSubcell.html">hpaVisSubcell</a></span><span class="op">(</span><span class="va">downloadedData</span>,</span>
<span>              targetGene<span class="op">=</span><span class="va">geneList</span>,</span>
<span>              customTheme<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">'Subcellular locations'</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">'Protein'</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle<span class="op">=</span><span class="fl">45</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span>  <span class="op">+</span></span>
<span>    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position<span class="op">=</span><span class="st">"none"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_equal</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="b_HPAanalyze_indepth_files/figure-html/visSubcell-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="individual-xml-import-and-image-downloading">Individual xml import and image downloading<a class="anchor" aria-label="anchor" href="#individual-xml-import-and-image-downloading"></a>
</h2>
<p>The <code>hpaXml</code> function family import and extract data from
individual XML entries from HPA. The <code><a href="../reference/hpaXmlGet.html">hpaXmlGet()</a></code> function
downloads and imports data as “xml_document”/“xml_node” object, which
can subsequently be processed by other <code>hpaXml</code> functions.
The XML format from HPA contains a wealth of information that may not be
covered by this package. However, users can extract any data of interest
from the imported XML file using the xml2 package.</p>
<p>A typical workflow for working with XML files includes the following
steps: (1) Download and import XML file with <code><a href="../reference/hpaXmlGet.html">hpaXmlGet()</a></code>;
(2) Extract the desired information with other <code>hpaXml</code>
functions; and (3) Download histological staining pictures, which is
currently supported by the <code>hpaXmlTissurExpr()</code> and
<code><a href="../reference/hpaXmlTissueExprSum.html">hpaXmlTissueExprSum()</a></code> functions.</p>
<div class="section level3">
<h3 id="the-umbrella-function-hpaxml">The umbrella function <code>hpaXml</code><a class="anchor" aria-label="anchor" href="#the-umbrella-function-hpaxml"></a>
</h3>
<p><code>hpaXml</code> will take an Ensembl gene id (start with
<em>ENSG</em>) and extract all availble information. You can also feed
the ourput of hpaXmlGet to it. See the quick-start vignette for more
details.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">EGFR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hpaXml.html">hpaXml</a></span><span class="op">(</span>inputXml<span class="op">=</span><span class="st">'ENSG00000146648'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">EGFR</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#&gt; [1] "ProtClass"     "TissueExprSum" "Antibody"      "TissueExpr"   </span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="import-xml-file-with-hpaxmlget">Import xml file with <code>hpaXmlGet()</code><a class="anchor" aria-label="anchor" href="#import-xml-file-with-hpaxmlget"></a>
</h3>
<p>The <code>hoaXmlGet()</code> function takes an Ensembl gene id (start
with <em>ENSG</em>) and import the perspective XML file into R. This
function calls the <code><a href="http://xml2.r-lib.org/reference/read_xml.html" class="external-link">xml2::read_xml()</a></code> under the hood, hence
the resulting object may be processed further with <em>xml2</em>
functions if desired.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">EGFRxml</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hpaXmlGet.html">hpaXmlGet</a></span><span class="op">(</span><span class="st">'ENSG00000146648'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="view-protein-classes-with-hpaxmlprotclass">View protein classes with <code>hpaXmlProtClass()</code><a class="anchor" aria-label="anchor" href="#view-protein-classes-with-hpaxmlprotclass"></a>
</h3>
<p>Protein class of queried protein can be extracted from the imported
XML with <code><a href="../reference/hpaXmlProtClass.html">hpaXmlProtClass()</a></code>. The output of this function is
a tibble of 4 columns: <code>id</code>, <code>name</code>,
<code>parent_id</code> and <code>source</code></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/hpaXmlProtClass.html">hpaXmlProtClass</a></span><span class="op">(</span><span class="va">EGFRxml</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#&gt; # A tibble: 40 x 4</span></span>
<span><span class="co">#&gt;    id    name                                   parent_id source    </span></span>
<span><span class="co">#&gt;    &lt;chr&gt; &lt;chr&gt;                                  &lt;chr&gt;     &lt;chr&gt;     </span></span>
<span><span class="co">#&gt;  1 Ez    Enzymes                                &lt;NA&gt;      &lt;NA&gt;      </span></span>
<span><span class="co">#&gt;  2 Ec    ENZYME proteins                        Ez        ENZYME    </span></span>
<span><span class="co">#&gt;  3 Et    Transferases                           Ec        ENZYME    </span></span>
<span><span class="co">#&gt;  4 Ki    Kinases                                Ez        UniProt   </span></span>
<span><span class="co">#&gt;  5 Kt    Tyr protein kinases                    Ki        UniProt   </span></span>
<span><span class="co">#&gt;  6 Ma    Predicted membrane proteins            &lt;NA&gt;      MDM       </span></span>
<span><span class="co">#&gt;  7 Md    Membrane proteins predicted by MDM     &lt;NA&gt;      MDM       </span></span>
<span><span class="co">#&gt;  8 Me    MEMSAT3 predicted membrane proteins    &lt;NA&gt;      MEMSAT3   </span></span>
<span><span class="co">#&gt;  9 Mf    MEMSAT-SVM predicted membrane proteins &lt;NA&gt;      MEMSAT-SVM</span></span>
<span><span class="co">#&gt; 10 Mg    Phobius predicted membrane proteins    &lt;NA&gt;      Phobius   </span></span>
<span><span class="co">#&gt; # ... with 30 more rows</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="get-summary-and-images-of-tissue-expression-with-hpaxmltissueexprsum">Get summary and images of tissue expression with
<code>hpaXmlTissueExprSum()</code><a class="anchor" aria-label="anchor" href="#get-summary-and-images-of-tissue-expression-with-hpaxmltissueexprsum"></a>
</h3>
<p>The function <code><a href="../reference/hpaXmlTissueExprSum.html">hpaXmlTissueExprSum()</a></code> extract the summary
of expression of protein of interest in normal tissue. The output of
this function is a list of (1) a string contains one-sentence summary
and (2) a dataframe of all tissues in which the protein was stained
positive and a histological stain images of those tissue.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/hpaXmlTissueExprSum.html">hpaXmlTissueExprSum</a></span><span class="op">(</span><span class="va">EGFRxml</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#&gt; $summary</span></span>
<span><span class="co">#&gt; [1] "Cytoplasmic and membranous expression in several tissues, most abundant in placenta."</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $img</span></span>
<span><span class="co">#&gt;            tissue</span></span>
<span><span class="co">#&gt; 1 cerebral cortex</span></span>
<span><span class="co">#&gt; 2      lymph node</span></span>
<span><span class="co">#&gt; 3           liver</span></span>
<span><span class="co">#&gt; 4           colon</span></span>
<span><span class="co">#&gt; 5          kidney</span></span>
<span><span class="co">#&gt; 6          testis</span></span>
<span><span class="co">#&gt; 7        placenta</span></span>
<span><span class="co">#&gt;                                                                imageUrl</span></span>
<span><span class="co">#&gt; 1 http://v18.proteinatlas.org/images/18530/41191_B_7_5_rna_selected.jpg</span></span>
<span><span class="co">#&gt; 2 http://v18.proteinatlas.org/images/18530/41191_A_7_8_rna_selected.jpg</span></span>
<span><span class="co">#&gt; 3 http://v18.proteinatlas.org/images/18530/41191_A_7_4_rna_selected.jpg</span></span>
<span><span class="co">#&gt; 4 http://v18.proteinatlas.org/images/18530/41191_A_9_3_rna_selected.jpg</span></span>
<span><span class="co">#&gt; 5 http://v18.proteinatlas.org/images/18530/41191_A_9_5_rna_selected.jpg</span></span>
<span><span class="co">#&gt; 6 http://v18.proteinatlas.org/images/18530/41191_A_6_6_rna_selected.jpg</span></span>
<span><span class="co">#&gt; 7 http://v18.proteinatlas.org/images/18530/41191_A_1_7_rna_selected.jpg</span></span></code></pre></div>
<p>Those images can be downloaded automatically by setting the
<code>downloadImg</code> argument to <code>TRUE</code>. Eg.
<code>hpaXmlTissueExprSum(CCNB1xml, downloadImg=TRUE)</code></p>
</div>
<div class="section level3">
<h3 id="get-details-of-individual-ihc-samples-with-hpaxmlantibody-and-hpaxmltissueexpr">Get details of individual IHC samples with
<code>hpaXmlAntibody()</code> and <code>hpaXmlTissueExpr()</code><a class="anchor" aria-label="anchor" href="#get-details-of-individual-ihc-samples-with-hpaxmlantibody-and-hpaxmltissueexpr"></a>
</h3>
<p>More importantly, the XML files are the only format of HPA
programmatically accesible data which contains information about each
antibody and each tissue sample used in the project.</p>
<p><code><a href="../reference/hpaXmlAntibody.html">hpaXmlAntibody()</a></code> extract the antibody information and
return a tibble with one row for each antibody.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/hpaXmlAntibody.html">hpaXmlAntibody</a></span><span class="op">(</span><span class="va">EGFRxml</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#&gt; # A tibble: 5 x 4</span></span>
<span><span class="co">#&gt;   id        releaseDate releaseVersion RRID      </span></span>
<span><span class="co">#&gt;   &lt;chr&gt;     &lt;chr&gt;       &lt;chr&gt;          &lt;chr&gt;     </span></span>
<span><span class="co">#&gt; 1 CAB000035 2006-03-13  1.2            &lt;NA&gt;      </span></span>
<span><span class="co">#&gt; 2 HPA001200 2008-02-15  3.1            AB_1078723</span></span>
<span><span class="co">#&gt; 3 HPA018530 2008-12-03  4.1            AB_1848044</span></span>
<span><span class="co">#&gt; 4 CAB068186 2014-11-06  13             AB_2665679</span></span>
<span><span class="co">#&gt; 5 CAB073534 2015-10-16  14             &lt;NA&gt;</span></span></code></pre></div>
<p><code><a href="../reference/hpaXmlTissueExpr.html">hpaXmlTissueExpr()</a></code> extract information about all samples
for each antibody above and return a list of tibbles. If antibody has
not been used for IHC staining, the returned tibble with be empty.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tissueExpression</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hpaXmlTissueExpr.html">hpaXmlTissueExpr</a></span><span class="op">(</span><span class="va">EGFRxml</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">tissueExpression</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#&gt;      Length Class  Mode</span></span>
<span><span class="co">#&gt; [1,] 18     tbl_df list</span></span>
<span><span class="co">#&gt; [2,] 18     tbl_df list</span></span>
<span><span class="co">#&gt; [3,] 18     tbl_df list</span></span>
<span><span class="co">#&gt; [4,] 18     tbl_df list</span></span>
<span><span class="co">#&gt; [5,] 18     tbl_df list</span></span></code></pre></div>
<p>Each tibble contain clinical data (<code>patientid</code>,
<code>age</code>, <code>sex</code>), tissue information
(<code>snomedCode</code>, <code>tissueDescription</code>), staining
results (<code>staining</code>, <code>intensity</code>,
<code>location</code>) and one <code>imageUrl</code> for each sample.
However, due to the large amount of data and the relatively large size
of each image, <code>hpaXmlTissueExpr</code> does not provide an
automated download option.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tissueExpression</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="co">#&gt; # A tibble: 327 x 18</span></span>
<span><span class="co">#&gt;    patientId age   sex   staining intensity quantity location imageUrl</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;     &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;    &lt;chr&gt;     &lt;chr&gt;    &lt;chr&gt;    &lt;chr&gt;   </span></span>
<span><span class="co">#&gt;  1 1653      53    Male  &lt;NA&gt;     &lt;NA&gt;      &lt;NA&gt;     &lt;NA&gt;     http://~</span></span>
<span><span class="co">#&gt;  2 1721      60    Fema~ &lt;NA&gt;     &lt;NA&gt;      &lt;NA&gt;     &lt;NA&gt;     http://~</span></span>
<span><span class="co">#&gt;  3 1725      57    Male  &lt;NA&gt;     &lt;NA&gt;      &lt;NA&gt;     &lt;NA&gt;     http://~</span></span>
<span><span class="co">#&gt;  4 4         25    Male  &lt;NA&gt;     &lt;NA&gt;      &lt;NA&gt;     &lt;NA&gt;     http://~</span></span>
<span><span class="co">#&gt;  5 512       34    Fema~ &lt;NA&gt;     &lt;NA&gt;      &lt;NA&gt;     &lt;NA&gt;     http://~</span></span>
<span><span class="co">#&gt;  6 2664      74    Fema~ &lt;NA&gt;     &lt;NA&gt;      &lt;NA&gt;     &lt;NA&gt;     http://~</span></span>
<span><span class="co">#&gt;  7 2665      88    Fema~ &lt;NA&gt;     &lt;NA&gt;      &lt;NA&gt;     &lt;NA&gt;     http://~</span></span>
<span><span class="co">#&gt;  8 1391      54    Fema~ &lt;NA&gt;     &lt;NA&gt;      &lt;NA&gt;     &lt;NA&gt;     http://~</span></span>
<span><span class="co">#&gt;  9 1447      45    Fema~ &lt;NA&gt;     &lt;NA&gt;      &lt;NA&gt;     &lt;NA&gt;     http://~</span></span>
<span><span class="co">#&gt; 10 1452      44    Fema~ &lt;NA&gt;     &lt;NA&gt;      &lt;NA&gt;     &lt;NA&gt;     http://~</span></span>
<span><span class="co">#&gt; # ... with 317 more rows, and 10 more variables: snomedCode1 &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   snomedCode2 &lt;chr&gt;, snomedCode3 &lt;chr&gt;, snomedCode4 &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   snomedCode5 &lt;chr&gt;, tissueDescription1 &lt;chr&gt;, tissueDescription2 &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   tissueDescription3 &lt;chr&gt;, tissueDescription4 &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   tissueDescription5 &lt;chr&gt;</span></span></code></pre></div>
<p><code>hpaTissueExprSum</code> and <code>hpaTissueExpr</code> provide
download links to download relevant staining images, with the former
function also gives the options to automate the downloading process.</p>
</div>
</div>
<div class="section level2">
<h2 id="compatibility-with-hpar-bioconductor-package">Compatibility with <code>hpar</code> Bioconductor package<a class="anchor" aria-label="anchor" href="#compatibility-with-hpar-bioconductor-package"></a>
</h2>
<table class="table">
<caption>(#tab:table) Complementary functionality between hpar and
HPAanalyze</caption>
<colgroup>
<col width="16%">
<col width="41%">
<col width="41%">
</colgroup>
<thead><tr class="header">
<th align="right">Functionality</th>
<th align="left">hpar</th>
<th align="left">HPAanalyze</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">Datasets</td>
<td align="left">Included in package</td>
<td align="left">Download from server or use built-in dataset</td>
</tr>
<tr class="even">
<td align="right">Query</td>
<td align="left">Ensembl id</td>
<td align="left">HGNC symbol for datasets, Ensembl id for XML</td>
</tr>
<tr class="odd">
<td align="right">Data version</td>
<td align="left">One stable version</td>
<td align="left">Latest by default, option to download older</td>
</tr>
<tr class="even">
<td align="right">Release info</td>
<td align="left">Access via functions</td>
<td align="left">N/A</td>
</tr>
<tr class="odd">
<td align="right">View relevant browser page</td>
<td align="left">Via <code>getHPA</code> function</td>
<td align="left">N/A</td>
</tr>
<tr class="even">
<td align="right">Visualization</td>
<td align="left">N/A</td>
<td align="left">Exploratory via <code>hpaVis</code> functions</td>
</tr>
<tr class="odd">
<td align="right">XML</td>
<td align="left">N/A</td>
<td align="left">Download and import via <code>hpaXml</code>
functions</td>
</tr>
<tr class="even">
<td align="right">Histology image</td>
<td align="left">View by loading browser page</td>
<td align="left">Extract links via <code>hpaXml</code> functions</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="acknowledgements">Acknowledgements<a class="anchor" aria-label="anchor" href="#acknowledgements"></a>
</h2>
<p>We appreciate the support of the National institutes of Health
National Cancer Institute R01 CA151522 and funds from the Department of
Cell, Developmental and Integrative Biology at the University of Alabama
at Birmingham.</p>
</div>
<div class="section level2">
<h2 id="copyright">Copyright<a class="anchor" aria-label="anchor" href="#copyright"></a>
</h2>
<p><strong>Anh Tran, 2018-2022 </strong></p>
<p>Please cite: <strong>Tran, A.N., Dussaq, A.M., Kennell, T. et
al. HPAanalyze: an R package that facilitates the retrieval and analysis
of the Human Protein Atlas data. BMC Bioinformatics 20, 463 (2019) <a href="https://doi.org/10.1186/s12859-019-3059-z" class="external-link uri">https://doi.org/10.1186/s12859-019-3059-z</a> </strong></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Anh Nhat Tran.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
